---
title: "Heart eQTLs"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source('../R/analysis_utils.R')
```


### Heart eQTL Finemapping

Here we focus on eQTLs that are finemapped with a posterior probability of association > 0.8. There are approximately 2,300 such SNPs.

```{r}
finemap.res <- suppressMessages(readr::read_tsv('../eQTL_enrich/gtex_finemapping/GTEx_v8_finemapping_DAPG/Heart_LV_Finemapping_CS95.txt', col_names = F))
colnames(finemap.res) <- c("tissue","gene","cluster_id","cluster_pip","variant_id","variant_pip")
```

```{r}
high_pip_eqtls <- unique(finemap.res$variant_id[finemap.res$variant_pip>0.8])
eqtl.gr <- snpIDtoGR(high_pip_eqtls)
eqtl.gr$SNP <- high_pip_eqtls
```

```{r}
rand.snps <- readLines('../matched_SNPs/eQTL_top_pip_hg19_5batches/snpsnap_match_hg38.txt')
rand.gr <- StringToGR(rand.snps)
rand.gr <- rand.gr[seqnames(rand.gr) %in% paste0("chr",1:22),]
seqlevels(rand.gr) <-  paste0("chr",1:22)
rand.gr$SNP <- paste0(seqnames(rand.gr),'_',start(rand.gr))
```

Get tissue activity of each eQTL from https://zenodo.org/record/3727189 (GTEx v8 paper)

```{r}
eqtl.lfsr <- suppressMessages(vroom::vroom('../eQTL_enrich/broadinstitute-gtex-v8-a014b43/data/Fig6C_all_top.z_lfsr.sig.pruned.txt.gz'))
```

Keep SNPs with PIP > 0.8 AND they have LFSR data.

```{r}
same.snps <- intersect(eqtl.gr$SNP, eqtl.lfsr$variant)
heart.eqtl.lfsr <- eqtl.lfsr[eqtl.lfsr$variant %in% same.snps,]
eqtl.gr <- eqtl.gr[eqtl.gr$SNP %in% same.snps]
```

Find the mean number of tissues per eQTL. 

```{r}
# all eqtls
tissues.active <- rowSums(eqtl.lfsr[,3:ncol(eqtl.lfsr)] < 0.01, na.rm = T) 
all.eqtl.ntissues <- data.frame(eqtl=eqtl.lfsr$variant, ntissues=tissues.active) %>% group_by(eqtl) %>% summarise(mean_tissues = mean(ntissues))
all.eqtl.ntissues <- all.eqtl.ntissues[all.eqtl.ntissues$mean_tissues > 0,]

# heart eqtls
tissues.active <- rowSums(heart.eqtl.lfsr[,3:ncol(heart.eqtl.lfsr)] < 0.01, na.rm = T) 
heart.eqtl.ntissues <- data.frame(eqtl=heart.eqtl.lfsr$variant, ntissues=tissues.active) %>% group_by(eqtl) %>% summarise(mean_tissues = mean(ntissues))
heart.eqtl.ntissues <- heart.eqtl.ntissues[heart.eqtl.ntissues$mean_tissues > 0,]
```

### Tissue Sharing of eQTLs

Below we plot the frequency of tissues active in causal heart eQTLs. Typically eQTLs have a bimodal distribution with peaks at 1-5 tissues and 45-50 tissues. We see that for heart eQTLs, they are generally shared across tissues. 

```{r}
betterHist <- function(X){
 bks <- seq(0, 50, 5)
 labs <- sapply(1:(length(bks)-1), function(x){paste0(bks[x]+1,'-',bks[x+1])})
 bin.count <- table(cut(X$mean_tissues, breaks = bks, labels = labs))
 bin.count <- bin.count/sum(bin.count)
 bin.count.df <- data.frame(count=as.numeric(bin.count), breaks=factor(names(bin.count), levels = labs))
}

all.eqtl.bin <- betterHist(all.eqtl.ntissues)
heart.eqtl.bin <- betterHist(heart.eqtl.ntissues)

result <- rbind(all.eqtl.bin, heart.eqtl.bin)
result$eQTLs <- c(rep(c("All","Heart"), each=nrow(all.eqtl.bin)))
```

```{r}
pdf('../manuscript_figures/figure4/Fig4A_tissue_sharing_eqtl.pdf', width=10, height=8)
ggplot(result, aes(x=breaks, y=count, fill=eQTLs)) + geom_bar(stat="identity", position = "dodge", width=0.8) + ggClean() + xlab('Tissues with LFSR < 0.01') + ylab('Proportion of eQTLs') + ggtitle('Tissue Sharing of Heart eQTLs') + theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + scale_fill_manual(values = c("#D6604D","#4393C3"))
dev.off()
```


Add tissue activity to our eQTL genomic ranges object

```{r}
eqtl.gr$ntissues.active <- heart.eqtl.ntissues$mean_tissues[match(eqtl.gr$SNP, heart.eqtl.ntissues$eqtl)]
#eqtl.gr.low <- eqtl.gr[which(eqtl.gr$ntissues.active<=10),]
#eqtl.gr.high <- eqtl.gr[which(eqtl.gr$ntissues.active>=30),]
```

### Distribution of heart eQTLs

Overlap generic genomic annotations with top eQTLs

```{r}
annots <- readRDS('../hg38_gtf_genomic_annots.gr.rds')
exon.annots <- annots$exons
utr.annots <- annots$UTRs
intron.annots <- annots$introns
```


Split ventrical cardiomyocyte peaks into cell-type specific and into shared.

```{r}
satac <- suppressMessages(ArchR::loadArchRProject('../ArchR/ArchR_heart_latest_noAtrium/', showLogo = F))
union.set <- ArchR::getPeakSet(satac)
union.exons <- union.set[union.set$peakType == "Exonic", ]

cmCalls <- readRDS('../ArchR/ArchR_heart_latest_noAtrium/PeakCalls/Cardiomyocyte-reproduciblePeaks.gr.rds')
cmCalls <- cmCalls[cmCalls$peakType != "Exonic",]

markers <- readRDS('../ArchR/ArchR_heart_latest_noAtrium/PeakCalls/DA_MARKERS_FDRP_1_log2FC_1.rds')
cm.specific <- markers$Cardiomyocyte
cm.specific <- removeOverlaps(X = cm.specific, to.remove = union.exons)

union.no.cm.specific <- removeOverlaps(X = union.set, to.remove = cm.specific)
non.cm.peaks <- removeOverlaps(X = union.no.cm.specific, to.remove = cmCalls)
cm.shared <- removeOverlaps(X = union.no.cm.specific, to.remove = non.cm.peaks)

intron.nonpeak <- removeOverlaps(intron.annots, to.remove = union.set)
```

Overlap eQTLs with exons, introns, etc.

```{r}

peaks.gr.list <- list("CM Specific" = cm.specific, 
                      "CM Shared" = cm.shared,
                      "Non-CM" = non.cm.peaks, 
                      "Intronic" = intron.nonpeak,
                      "Exon" = exon.annots, 
                      "UTR" = utr.annots)

peaks.eqtls <- join_overlap_list(gr.list = peaks.gr.list, X = eqtl.gr)
peaks.random <- join_overlap_list(gr.list = peaks.gr.list, X = rand.gr)
```

```{r}
snpsIn <- unique(
  unlist(
    sapply(peaks.eqtls, function(x){x$SNP})
  )
)
peaks.eqtls$Unassigned <- eqtl.gr[!(eqtl.gr$SNP %in% snpsIn),]

snpsIn <- unique(
  unlist(
    sapply(peaks.random, function(x){x$SNP})
  )
)
peaks.random$Unassigned <- rand.gr[!(rand.gr$SNP %in% snpsIn),]
```


```{r}
peak.dist.df <- as.data.frame(sapply(peaks.eqtls, FUN = function(x){length(x)/length(eqtl.gr)}))
colnames(peak.dist.df) <- c("freq")
peak.dist.df$category <- rownames(peak.dist.df)

random.dist.df <- as.data.frame(sapply(peaks.random, FUN = function(x){length(x)/length(rand.gr)}))
colnames(random.dist.df) <- c("freq")
random.dist.df$category <- rownames(random.dist.df)
```

```{r}
peak.set.dist.df <- Reduce(rbind, list(random.dist.df, peak.dist.df))
peak.set.dist.df$SNPs <- c(rep("Random SNPs", nrow(random.dist.df)),
                          rep("eQTLs", nrow(peak.dist.df)))
```

```{r}
plot.levels <- c("Exon","UTR","Intronic","CM Specific","CM Shared", "Non-CM", "Unassigned")
peak.set.dist.df$category <- factor(peak.set.dist.df$category, 
                                    levels = rev(plot.levels))

pdf('../manuscript_figures/figure4/Fig4B_wc_distribution_1.pdf', width=10, height=8)
ggplot(peak.set.dist.df, aes(y=category, x=freq, fill=SNPs)) + 
  geom_bar(stat="identity", position="dodge") + ggClean() + xlab("Wc (prop. of SNPs)") + ylab("Category") + ggtitle("Distribution of Top Heart eQTLs") + scale_fill_manual(values = c("black","gray"))  + coord_cartesian(xlim=c(0, 0.4))
dev.off()
```


Tissue sharing patterns in scATAC-seq peak sets

```{r}
peaks.tissue.sharing <- sapply(peaks.eqtls, function(x){x$ntissues.active}) 
peaks.tissue.sharing$`Union Set` <- NULL
peaks.tissue.sharing <- setNames(unlist(peaks.tissue.sharing, use.names=F),rep(names(peaks.tissue.sharing), lengths(peaks.tissue.sharing)))
peaks.tissue.sharing.df <- data.frame(category=names(peaks.tissue.sharing), ntissues=peaks.tissue.sharing)
peaks.tissue.sharing.df$category <- factor(peaks.tissue.sharing.df$category, levels=rev(plot.levels))
peaks.tissue.sharing.df <- peaks.tissue.sharing.df[!is.na(peaks.tissue.sharing.df$ntissues),]
```

```{r}
pc_shared <- peaks.tissue.sharing.df %>% group_by(category) %>% summarise(p_shared = mean(ntissues > 15, na.rm = T))

pdf("../manuscript_figures/figure4/Fig4B_prob_sharing.pdf", width=6, height=8)
ggplot(pc_shared, aes(y = category, x = p_shared)) + geom_bar(stat="identity", fill="black") + ggClean() + xlab("Pc (prob. sharing)") + ylab("Category") + xlim(c(0,1))
dev.off()
```

```{r}
pdf("../manuscript_figures/figure4/Fig4B_tissues_active_dist.pdf", width=6, height=8)

ggplot(peaks.tissue.sharing.df, aes(y=category, x=ntissues)) + 
  geom_violin(fill="black") + 
  geom_boxplot(width=0.1, color="lightblue") + 
  ggClean() + LegendOff() + xlab("Tissues Active (LFSR < 0.01)") + ylab("") +
  scale_x_continuous(breaks=seq(0,50,10))

dev.off()
```

Expression patterns

```{r}
high_pip_genes <- finemap.res$gene[finemap.res$variant_id %in% peaks.eqtls$`CM Specific`$SNP]
high_pip_genes <- sub('[.].*', '', high_pip_genes)
high_pip_gene_symbol <- ensembldb::select(EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86, keys= high_pip_genes, keytype = "GENEID", columns = "SYMBOL")
high_pip_gene_symbol <- high_pip_gene_symbol$SYMBOL
```

```{r}
srna <- readRDS('../seurat/Heart_RNA_Processed_Combined_NoAtrium.rds')
```

```{r}
get_cell_type_expr <- function(srna, genes){
  rna.mat <- srna@assays$RNA@data
  cm.specific.exp <- rna.mat[rownames(rna.mat) %in% genes,]
  celltypes <- unique(Seurat::Idents(srna))
  mean.exp <- list()
  for(t in celltypes){
    mean.exp[[t]] <- rowMeans(cm.specific.exp[,Seurat::Idents(srna) == t])
  }
  exp.mat <- as.data.frame(mean.exp)
  exp.mat <- as.matrix(exp.mat)
  exp.mat.z <- sweep(exp.mat - rowMeans(exp.mat), MARGIN = 1, STAT = matrixStats::rowSds(exp.mat), FUN = '/') 
  
  return(exp.mat.z)
}
```

```{r}
CM.specific.egenes.exp <- get_cell_type_expr(srna, high_pip_gene_symbol)
```

```{r}
high_pip_genes <- finemap.res$gene[finemap.res$variant_id %in% peaks.eqtls$`CM Shared`$SNP]
high_pip_genes <- sub('[.].*', '', high_pip_genes)
high_pip_gene_symbol <- ensembldb::select(EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86, keys= high_pip_genes, keytype = "GENEID", columns = "SYMBOL")
high_pip_gene_symbol <- high_pip_gene_symbol$SYMBOL
```

```{r}
CM.shared.egenes.exp <- get_cell_type_expr(srna, high_pip_gene_symbol)
```

```{r}
CM.specific.egenes.exp <- reshape2::melt(CM.specific.egenes.exp)
CM.shared.egenes.exp <- reshape2::melt(CM.shared.egenes.exp)
CM.egenes.exp <- rbind(CM.specific.egenes.exp, CM.shared.egenes.exp)
CM.egenes.exp[,"type"] <- c(rep("CM Specific eGenes", nrow(CM.specific.egenes.exp)), rep("CM Shared eGenes", nrow(CM.shared.egenes.exp)))
```

```{r}
ggplot(CM.egenes.exp, aes(x=Var2, y=value, fill=type)) + geom_boxplot() + ggClean(rotate_axis = T) + xlab("") + ylab("RNA Expression")
```


```{r}

eGene_Enrichment <- function(eGenes, seurat.rna, rna.markers){
  
  cell.types <- as.character(unique(rna.markers$cluster))
  result <- list()
  
  for( ct in c("Vent. CM")){
    
    rna.markers.filt <- rna.markers[rna.markers$avg_logFC > 0.1 & rna.markers$p_val_adj < 2e-06,,]
    curr.genes <- row.names(rna.markers.filt)
    print(nrow(rna.markers.filt))
    
    N <- sum(Idents(seurat.rna) == ct)
    non.specific <- row.names(rna.markers)[!(rownames(rna.markers) %in% curr.genes)]
    curr.exp <- (seurat.rna@assays$RNA@counts > 0)
    curr.exp <- rowSums(curr.exp[non.specific, Idents(seurat.rna) == ct]) / N
    shared.genes <- non.specific[curr.exp > 0.20]

    genes.eqtls <- intersect(eGenes, curr.genes)
    shared.genes.eqtls <- intersect(eGenes, shared.genes)

    result[[ct]] <- data.frame(type = c("Specific", "Shared"),
                               pval = c(1 - phyper(q = length(genes.eqtls) - 1, m = length(curr.genes), n = 19429-length(curr.genes), k = length(eGenes)),
                                        1 - phyper(q = length(shared.genes.eqtls) - 1, m = length(shared.genes), n = 19429-length(shared.genes), k = length(eGenes))),
                               overlap = c(length(genes.eqtls)/length(eGenes), length(shared.genes.eqtls)/length(eGenes))
                               )
    
  }
  return(result)
}

```

```{r}
rna.markers <- readRDS('../seurat/CM_rna_markers.rds')
```

```{r}
high_pip_genes <- finemap.res$gene[finemap.res$variant_id %in% peaks.eqtls$`CM Specific`$SNP & finemap.res$variant_pip > 0.8]
high_pip_genes <- unique(sub('[.].*', '', high_pip_genes))
high_pip_gene_symbol <- ensembldb::select(EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86, keys= high_pip_genes, keytype = "GENEID", columns = "SYMBOL")$SYMBOL

specific.enrich <- eGene_Enrichment(eGenes = high_pip_gene_symbol, seurat.rna = srna, rna.markers = rna.markers)
```

```{r}
high_pip_genes <- finemap.res$gene[finemap.res$variant_id %in% peaks.eqtls$`CM Shared`$SNP & finemap.res$variant_pip > 0.8]
high_pip_genes <- unique(sub('[.].*', '', high_pip_genes))
high_pip_gene_symbol <- ensembldb::select(EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86, keys= high_pip_genes, keytype = "GENEID", columns = "SYMBOL")$SYMBOL

shared.enrich <- eGene_Enrichment(eGenes = high_pip_gene_symbol, seurat.rna = srna, rna.markers = rna.markers)
```

```{r}
high_pip_genes <- finemap.res$gene[finemap.res$variant_pip > 0.8]
high_pip_genes <- unique(sub('[.].*', '', high_pip_genes))
high_pip_gene_symbol <- ensembldb::select(EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86, keys= high_pip_genes, keytype = "GENEID", columns = "SYMBOL")$SYMBOL

all.enrich <- eGene_Enrichment(eGenes = high_pip_gene_symbol, seurat.rna = srna, rna.markers = rna.markers)
```

```{r}
random.genes <- rownames(srna)[sample(1:nrow(srna), size = 1000, replace = F)]
rand.enrich <- eGene_Enrichment(eGenes = high_pip_gene_symbol, seurat.rna = srna, rna.markers = rna.markers)
```

```{r}
enrich.df <- rbind(specific.enrich$`Vent. CM`, shared.enrich$`Vent. CM`, all.enrich$`Vent. CM`, rand.enrich$`Vent. CM`)
enrich.df["mlog10pval"] <- -log10(enrich.df$pval)
enrich.df["eGenes"] <- c(rep("CM Specific eGenes", 2), rep("CM Shared eGenes", 2), rep("All eGenes (PIP>0.8)", 2), rep("Random", 2))
```

```{r}
p1 <- ggplot(enrich.df, aes(x=type, y=overlap, fill=eGenes)) + geom_bar(stat="identity", position="dodge") + ggClean(rotate_axis = T) + xlab("Cardiomyocyte Genes") + ylab("Proportion of eGenes") 
```

```{r}
p2 <- ggplot(enrich.df, aes(x=type, y=mlog10pval, fill=eGenes)) + geom_bar(stat="identity", position="dodge") + ggClean(rotate_axis = T) + xlab("Cardiomyocyte Genes") + ylab("-log10 pval") + LegendOff()
```

```{r}
p1 + p2
```

