---
title: "Finemapping Results"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(liftOver)
source('../R/analysis_utils.R')
source('../R/track_plotter.R')
```

### Enrichment of various GWAS traits

I ran `R/torus_enrichment.R` to obtain the Torus enrichment results, which we visualize below. 

```{r}
torus.enrichment <- readRDS('../GWAS/Torus_Enrichment_Results_Univariate.df.rds')

res <- lapply(torus.enrichment, function(x){ Reduce(x = x, f = rbind)})
res <- lapply(res, function(x){x[x$term != "Intercept",]})
estimates <- as.data.frame(sapply(res, function(x){x["estimate"]}))
low <- as.data.frame(sapply(res, function(x){x[,"low",drop=F]}))
```

```{r}
rnames <- c("Heart eQTLs","CM Shared ATAC","CM Specific ATAC","Coding","Conserved","Non-CM ATAC")
row.names(estimates) <- rnames
row.names(low) <- rnames
colnames(estimates) <- sub(".estimate","", colnames(estimates))
colnames(low) <- sub(".estimate","", colnames(estimates))
```

```{r}
ComplexHeatmap::Heatmap(t(estimates), cluster_rows = T, cluster_columns = F, name = "Torus Estimate", show_row_dend = F)
```

### Finemapping results

I ran the script `R/run_finemapping.R` on each locus individually using `broadwl` for maximum parallelization. Below we load the results and concatenate. `finemappeR` was used to add prior probability to the GWAS summary statistics. 

```{r}
suppressMessages(library(tidyverse))
suppressMessages(library(finemappeR))

concat.fm.files <- function(file.names){
  
  stopifnot(length(file.names) > 1)
  rds.list <- list()
  for(f in file.names){
    rds.list[[f]] <- readRDS(f)
  }
  
  res <- Reduce(x = rds.list, f = rbind)
  
  res %>% as_tibble()
  
}


prior.files <- list.files(path = '../GWAS/finemapping/locus_files_finemapping', pattern = 'Finemapped_Prior*', full.names = T)
uniform.files <- list.files(path = '../GWAS/finemapping/locus_files_finemapping', pattern = 'Finemapped_Uniform*', full.names = T)


prior_res <- concat.fm.files(prior.files)
unif_res <- concat.fm.files(uniform.files)
```

```{r, eval=F}
saveRDS(prior_res, '../GWAS/finemapping/aFib_Finemapped.tble.rds')
```


```{r}
finemap.CS.size <- prior_res %>% group_by(locus) %>% filter(CS == 1) %>% summarise(n = dplyr::n()) %>% .$n
```

```{r, fig.width=16, fig.height=2}
bks <- c(0, 1, 5, 10, 20, 50, 10000)
labs <- c("1", "2-5","6-10","11-20","21-50", "51+")
bin.count <- table(cut(finemap.CS.size, breaks = bks, labels = labs))
bin.count.df <- data.frame(count=as.numeric(bin.count), Size=factor(names(bin.count), levels = rev(labs)), group = "A", interval_size = c(2, 4, 4, 9, 29, 10))

ggplot(bin.count.df, aes(x = group, y=count, fill=Size)) + 
  geom_col() + 
  ggClean() + 
  xlab('Genetic Credible Set Size') + 
  ylab('Frequency') + 
  ggtitle('Number of causal variants / Number of loci') + 
  theme_void() +
  coord_flip() +
  geom_text(aes(label = paste0("\n", count)), color="white", position = position_stack(vjust = 0.5)) +
  geom_text(aes(label = paste0(Size, "\n")), position = position_stack(vjust = 0.5)) +
  theme(legend.position = 'none') +
  scale_fill_manual(values = c("#aaaaaa",brewer.pal(n = 6, name = "Greens")[-1]))
 
```

```{r, fig.width=16, fig.height=2}
bks <- c(0.01, 0.1, 0.5, 0.8, 0.9, 0.95, 0.99, 1)
labs <- c("1-10%","10-50%","50-80%","80-90%","90-95%","95-99%",">99%")
bin.count <- table(cut(prior_res$susie_pip, breaks = bks, labels = labs))
bin.count.df <- data.frame(count=as.numeric(bin.count), Size=factor(names(bin.count), levels = labs), group = "A", interval_size = c(10, 40, 30, 10, 6, 6, 4))

ggplot(bin.count.df, aes(x = group, y=interval_size, fill=Size)) + 
  geom_col() + 
  ggClean() + 
  xlab('Genetic Credible Set Size') + 
  ylab('Frequency') + 
  ggtitle('Posterior probability of being a causal variant') + 
  theme_void() +
  coord_flip() +
  geom_text(aes(label = paste0("\n", count)), color="white", position = position_stack(vjust = 0.5)) +
  geom_text(aes(label = paste0(Size, "\n")), position = position_stack(vjust = 0.5)) +
  theme(legend.position = 'none') +
  scale_fill_manual(values = c("#aaaaaa",brewer.pal(n = 7, name = "Greens")[-1]))


```


```{r}
sumstat.annot <- annotator(gwas = prior_res, annotations = list.files('../GWAS/annotations_for_finemapping_hg19/', '*.bed', full.names = T))
annots <- rep("Other", nrow(sumstat.annot))
annots[sumstat.annot$CM_shared_peaks_hg19.bed_d>0] <- "CM Shared ATAC"
annots[sumstat.annot$CM_specific_peaks_hg19.bed_d>0] <- "CM Specific ATAC"
annots[sumstat.annot$non_CM_peaks_hg19.bed_d>0] <- "Non-CM ATAC"
annots[sumstat.annot$all_eqtls_hg19.bed_d>0] <- "Other" #CHANGE BACK to eQTLs
```

```{r}
annot.snp.df <- data.frame(snp=prior_res$snp,
                           prior=prior_res$susie_pip, 
                          uniform=unif_res$susie_pip, 
                          annots=annots)
```

Here we visualize the effect of genomic annotations on the PIP of each SNP

```{r}
ggplot(annot.snp.df) + 
  geom_point(aes(x=uniform, y = prior, color = annots)) + 
  geom_point(data = subset(annot.snp.df, annots=="eQTLs"), aes(x=uniform, y = prior, color = annots)) +
  xlab('PIP (uniform prior)') + ylab('PIP (annotation prior)') + ggClean() +
  scale_color_manual(values = c("Red","Blue","Green","black","grey"))
```

### Linking causal SNPs to Genes

Filter finemapping results

```{r}
finemap.gr <- GenomicRanges::GRanges(seqnames = prior_res$chr, ranges = IRanges::IRanges(start = prior_res$pos, end = prior_res$pos))
finemap.gr$snp <- prior_res$snp
finemap.gr$pip <- prior_res$susie_pip
finemap.gr$chr <- prior_res$chr
finemap.gr$pos <- prior_res$pos
finemap.gr$locus <- prior_res$locus
seqlevelsStyle(finemap.gr) <- "UCSC"
finemap.gr <- finemap.gr[finemap.gr$pip > 1e-5, ]
```

load annotations for linking SNPs to genes (run `clean_genomic.annotations.R` to produce)

```{r}
genomic.annots <- readRDS('../hg19_gtf_genomic_annots.gr.rds')
gene.annots <- genomic.annots$genes
genomic.annots$genes <- NULL
```

Download Hi-C data from https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-6014/ and convert the dataframe into a GRanges object.

```{r}
cm.hic <- suppressMessages(read_tsv('../GWAS/HiC/CM+iPSC_Hi-C/capt-CHiCAGO_interactions-CM-1.ibed')) #hg19

bait.gene.names <- sub(pattern = "\\*.*", replacement = "" , cm.hic$bait_name)
cm.hic.enhancer.gr <- GRanges(seqnames = cm.hic$otherEnd_chr, IRanges(start = cm.hic$otherEnd_start, end = cm.hic$otherEnd_end)) 
cm.hic.enhancer.gr$promoter_chr <- cm.hic$bait_chr
cm.hic.enhancer.gr$promoter_start <- cm.hic$bait_start
cm.hic.enhancer.gr$promoter_end <- cm.hic$bait_end
cm.hic.enhancer.gr$gene_name <- bait.gene.names

cm.hic.enhancer.gr <- cm.hic.enhancer.gr[cm.hic.enhancer.gr$gene_name %in% gene.annots$gene_name,] # restrict to protein coding genes
genomic.annots$HiC <- cm.hic.enhancer.gr
```

Add co-accessibility by running the script `R/prepare_Coaccessibility.R` and load it here

```{r}
coaccess <- readRDS('../ArchR/ArchR_heart_latest_noAtrium/Coaccess_enhancers_hg19.gr.rds')
coaccess <- coaccess[coaccess$gene_name %in% gene.annots$gene_name,]
genomic.annots$coacc <- coaccess
```

Overlap finemapped SNPs with each of our annotations

```{r}
snps.overlap.annots <- GRangesList(lapply(genomic.annots, function(x) { plyranges::join_overlap_inner(x = x, y = finemap.gr) }))
snps.in <- unique(unlist(snps.overlap.annots)$snp)
snps.unassigned.gr <- finemap.gr[!(finemap.gr$snp %in% snps.in),]

res <- plyranges::join_overlap_inner(x = genomic.annots$promoters, 
                                     y = snps.unassigned.gr, maxgap = 1e6) # get all promoters within 1MB

res$distance <- abs(res$pos - end(res))
res$weight <- exp(-res$distance / 1e5)

snps.overlap.annots$intronic <- res
```

```{r}
categories <- names(snps.overlap.annots)
mat.list <- list()

for(c in categories){
  curr <- snps.overlap.annots[[c]]@elementMetadata %>% as_tibble()
  if(c == "intronic"){
    curr <- curr[,c("snp","chr", "pos","locus","pip","gene_name", "weight")] %>% mutate(category = "distance")
  }
  else{
    curr <- curr[,c("snp","chr", "pos","locus","pip","gene_name", "weight")] %>% mutate(category = c, weight = 1*(pip>0))
  }
  mat.list[[c]] <- curr
} 

final.mat <- Reduce(rbind, mat.list) %>% distinct(gene_name, snp, pip, .keep_all = TRUE)
```

```{r}
final.mat <- final.mat %>% group_by(snp) %>% mutate(frac_pip = weight/sum(weight))
final.mat <- final.mat %>% group_by(gene_name) %>% mutate(gene_pip = sum(pip * frac_pip)) %>% arrange(gene_name)

snp.annots <- data.frame(snp=prior_res$snp, annots=annots)
final.mat <- dplyr::left_join(x = final.mat, y = snp.annots, by = "snp")
final.mat <- final.mat[!is.na(final.mat$gene_name),]
```

```{r}
saveRDS(final.mat, '../GWAS/finemapping/aFib_Finemapped_GeneMapped.tble.rds')
```

### prepare tables for experimental verification

```{r}
gene.view.df <- final.mat %>% dplyr::select(locus, gene_name, gene_pip, pip) %>% group_by(locus, gene_name) %>% summarise(gene_pip = round(gene_pip[1], 3), n_snps = sum(pip > 0.01))
high.conf.snp.df <- final.mat %>% dplyr::filter(pip > 0.2) %>% group_by(snp) %>% arrange(-gene_pip) %>% dplyr::slice(1) 
```

```{r}
snp.gene <- high.conf.snp.df %>% dplyr::select(snp, pos, gene_name)
gene.locs.df <- gene.annots %>% as_tibble()

snp.gene.dist <- snp.gene %>% left_join(., gene.locs.df, on = 'gene_name') %>%
    mutate(dist = abs(start - pos)) %>%
    dplyr::select(snp, gene_name, dist)

high.conf.snp.df <- inner_join(high.conf.snp.df, snp.gene.dist, on = c('snp','gene_name'))

```

I downloaded ALL H3K27ac bed files from Encode for Aorta, Heart LV/RV/RA, and cardiac muscle. I concatenated them into one bed file. They were in Hg38 so I converted them to Hg19 with liftOver. 

```{r}
seqlevelsStyle(finemap.gr) <- "NCBI"
files <- list.files(path = '../ENCODE/H3k27ac/hg19_mapped', pattern = '*.bed', full.names = T)
files <- c(files, '../ENCODE/DNase/FetalHeart_E083-DNase_hg19_cleaned_narrowPeak.bed.gz')
annotated.snps <- lapply(files, function(x){get_elements_overlap_snps(snp.gr = finemap.gr, annotations = x)})
```

```{r}
for(i in 1:length(annotated.snps)){
    high.conf.snp.df <- left_join(high.conf.snp.df, annotated.snps[[i]], on = 'snp')
}
```

SNP-level table at >20% PIP

```{r}
high.conf.snp.df %>% dplyr::select(-weight, -frac_pip, -locus) %>% dplyr::rename(SNP = snp, `b37 bp` = pos, PIP = pip, `Gene Linked` = gene_name, `Gene PIP`=gene_pip, `Link Method`=category, `Distance to Gene`=dist, `Chromatin status`=annots) %>% mutate(PIP = round(PIP, 3), `Gene PIP` = round(`Gene PIP`, 3)) %>% write_csv('../GWAS/aFib_Finemapped_minPIP_0.2_03252021.csv')
```


```{r}

nielsen_snp_genes <- suppressMessages(read_csv('../GWAS/Nielsen_snps_prior_genes.csv', col_names = T)) %>% dplyr::rename(snp = rsID)
snp.pos <- as.integer(sub("chr.*:","",nielsen_snp_genes$`Position (hg19)`))
top.snp.gr <- GRanges(seqnames = sub(":.*","",nielsen_snp_genes$`Position (hg19)`), ranges = IRanges(start = snp.pos, end = snp.pos)) + 1e6
top.snp.gr$snp <- nielsen_snp_genes$snp
top.snp.gr$nielsen_locus <- nielsen_snp_genes$`Locus No.`
top.snp.gr$genes_prioritized <- nielsen_snp_genes$`Prioritized genes`

data('Euro_LD_Chunks', package='finemappeR')
ldblocks.gr <- GRanges(seqnames = paste0("chr",LD_Blocks$X1), ranges = IRanges(start = LD_Blocks$X2, end = LD_Blocks$X3), locus=LD_Blocks$X4)

locus.convert <- plyranges::join_overlap_inner(top.snp.gr, ldblocks.gr) %>% as_tibble() %>% dplyr::select(nielsen_locus, locus, genes_prioritized)

```

Locus/gene view

```{r}
gene.cumsum.df <- gene.view.df %>% 
    group_by(locus) %>% 
    arrange(-gene_pip) %>% mutate(gene_pip_csum = cumsum(gene_pip)) %>% 
    dplyr::slice(1:which(gene_pip_csum >= 0.8)[1]) 

gene.cs.df <- gene.cumsum.df %>% group_by(locus) %>% summarise(gene_cs = paste0(gene_name, collapse=','), top_gene = gene_name[1], top_gene_pip = gene_pip[1])

gene_view_df <- left_join(gene.cumsum.df, locus.convert, by = 'locus') %>% 
    ungroup() %>%
    dplyr::rename(Locus = nielsen_locus, `Gene Name` = gene_name, `Gene PIP` = gene_pip, `N SNPs Linked` = n_snps) %>%
    dplyr::select(Locus, `Gene Name`, `Gene PIP`, `N SNPs Linked`) %>%
    write_csv('../GWAS/aFib_Finemapped_GeneView.csv')

locus_view_df <- left_join(gene.cs.df, locus.convert, by = 'locus') %>% 
    group_by(nielsen_locus) %>% 
    summarise(`Nielsen Gene` = paste0(unique(genes_prioritized), collapse=','),
              `80% Gene Credible Set` =  paste0(unique(gene_cs), collapse=','),
              `Top Genes` = paste0(unique(top_gene), collapse=','),
              `Top Gene PIP` = paste0(unique(top_gene_pip), collapse=',')) %>%
    dplyr::rename(Locus = nielsen_locus) %>%
    write_csv('../GWAS/aFib_Finemapped_LocusView.csv')
```



```{r}
final.mat.high.conf <- left_join(final.mat.high.conf, nielsen_snp_genes, by = 'snp') 

final.mat.high.conf[is.na(final.mat.high.conf)] <- ""
final.mat.high.conf$is_prioritized <- ifelse(yes = "Yes", no = "No", test = (final.mat.high.conf$gene_name  == final.mat.high.conf$`Prioritized genes`))
```

```{r}
final.mat.high.conf %>% dplyr::select(-weight, -frac_pip) %>% arrange(chr, pos, -pip) %>% write_csv('../GWAS/aFib_Finemapped_MinPip0.2_03232021.csv')
```


### Gene Credible Sets

code below is wrong, fix edge case when 1 SNP has a PIP higher than the CS limit
```{r}
cs.thresh <- 0.9
finemap.res <- suppressMessages(read_csv('../GWAS/gene_pips_summary.csv'))
gene.cs <- finemap.res %>% 
  dplyr::select(c(gene_name,gene_pip,locus)) %>% 
  distinct() %>% 
  group_by(locus) %>% 
  arrange(desc(gene_pip)) %>% 
  mutate(gene_pip_cumsum = cumsum(gene_pip)) %>%
  filter(gene_pip_cumsum <= 0.9)
```

```{r, fig.width=14, fig.height=2}
gene.cs.size <- gene.cs %>% group_by(locus) %>% summarise(n = dplyr::n()) %>% .$n
```

```{r, fig.width=16, fig.height=2}
bks <- c(0, 1, 5, 10, 20, 100)
labs <- c("1", "2-5","6-10","11-20","21+")
bin.count <- table(cut(gene.cs.size, breaks = bks, labels = labs))
bin.count.df <- data.frame(count=as.numeric(bin.count), Size=factor(names(bin.count), levels = rev(labs)), group = "A")

ggplot(bin.count.df, aes(x = group, y=count, fill=Size)) + 
  geom_col() + 
  ggCleadplyr::n() + 
  xlab('Genetic Credible Set Size') + 
  ylab('Frequency') + 
  ggtitle('Number of causal genes / Number of loci') + 
  theme_void() +
  coord_flip() +
  geom_text(aes(label = paste0("\n", count)), color="white", position = position_stack(vjust = 0.5)) +
  geom_text(aes(label = paste0(Size, "\n")), position = position_stack(vjust = 0.5)) +
  theme(legend.position = 'none') +
  scale_fill_manual(values = c("#aaaaaa",brewer.pal(n = 6, name = "Greens")[-1]))
 
```

### Track plotting 

Add LD information to our finemapped results. Move this to a script...

```{r}
big.snp <- bigsnpr::snp_attach('/project2/xinhe/1kg/bigsnpr/EUR_variable_1kg.rds')
```


```{r}
# camk2d locus : 472
# akap5 locus : 1352
pip.df <- prior_res[prior_res$locus==1414 | prior_res$locus==1413,] 

s <- min(pip.df$pos)
e <- max(pip.df$pos)
snp.p <- pip.df$pos[pip.df$snp == pip.df$snp[which.max(pip.df$susie_pip)] ]
ss <- max(s, snp.p - 1e5)
ee <- min(e, snp.p + 1e5)
curr.locus.gr <- GRanges(seqnames = paste0("chr",pip.df$chr[1]), IRanges(start =  ss, 
                                                                         end = ee ))
pip.df$isAnnotate <- pip.df$susie_pip > 0.1
top.snp <- pip.df$bigSNP_index[which.max(pip.df$pval)]

top.snp.G <- big.snp$genotypes[,top.snp]
G.mat <- big.snp$genotypes[,pip.df$bigSNP_index]
r2.vals <- as.vector(cor(top.snp.G, G.mat))^2
r2.brackets <- cut(r2.vals, breaks = c(0,0.1, 0.25, 0.75, 0.9, 1), labels = c("0-0.1","0.1-0.25","0.25-0.75","0.75-0.9","0.9-1"))
pip.df$r2 <- r2.brackets
```

Prepare HiC dataframe

```{r}
cm.hic.enhancer.gr.snps <- plyranges::join_overlap_left(cm.hic.enhancer.gr, finemap.gr) %>% dplyr::select(-chr, -pos, -pip)

hicplot.df <- cm.hic.enhancer.gr.snps %>% 
  as_tibble() %>% 
  mutate( enhancer_mid = (start+end)/2,
          promoter_mid = (promoter_start + promoter_end)/2) %>% 
  dplyr::rename(chr = seqnames)
```

Prepare coacc dataframe

```{r}
other.ranges.clean.hg19 <- plyranges::join_overlap_left(other.ranges.clean.hg19, finemap.gr %>% dplyr::select(-chr)) %>% dplyr::select(-pos, -pip)

x <- other.ranges.clean.hg19 %>% as_tibble() %>% dplyr::select(-seqnames, -start, -end, -width, -strand)
y <- promoter.ranges.clean.hg19 %>% as_tibble() %>% dplyr::select(-seqnames, -start, -end, -width, -strand)

coaccplot.df <- inner_join(x, y, by = 'idx') %>% mutate(enhancer_mid = (enhancer_end+enhancer_start)/2, promoter_mid = (promoter_start+promoter_end)/2)
```


Read in bedGraph files for ATAC-seq peaks

```{r}
cm.atac.bed <- rtracklayer::import('../ArchR/ArchR_heart_latest/GroupBigWigs/CellTypes/Hg19_Cardiomyocyte-TileSize-100-normMethod-ReadsInTSS-ArchR.bw.bedGraph')
fib.atac.bed <- rtracklayer::import('../ArchR/ArchR_heart_latest/GroupBigWigs/CellTypes/Hg19_Fibroblast-TileSize-100-normMethod-ReadsInTSS-ArchR.bw.bedGraph')
lymph.atac.bed <- rtracklayer::import('../ArchR/ArchR_heart_latest/GroupBigWigs/CellTypes/Hg19_Lymphoid-TileSize-100-normMethod-ReadsInTSS-ArchR.bw.bedGraph')
```


```{r}
plot.list <- list()
plot.list[["pval.plot"]] <- plot.pval.track(pip.df = pip.df, curr.locus.gr = curr.locus.gr)
plot.list[["pip.plot"]] <- plot.pip.track(pip.df = pip.df, curr.locus.gr = curr.locus.gr)
plot.list[["gtrack"]] <- geneTrackPlot(curr.locus.gr = curr.locus.gr, collapseTranscripts = "longest")

plot.list[["cm.track"]]  <- plotATAC(cm.atac.bed, region.focus = curr.locus.gr)
plot.list[["fib.track"]]  <- plotATAC(fib.atac.bed, region.focus = curr.locus.gr)
plot.list[["lymph.track"]]  <- plotATAC(lymph.atac.bed, region.focus = curr.locus.gr)

# plot.list[["hic"]] <- HiC.track(HiC.Midpoints = hicplot.df, 
#                                 curr.locus.gr = curr.locus.gr, 
#                                 links.to.highlight = hicplot.df$snp %in% pip.df$snp[pip.df$isAnnotate==T])

plot.list[["hic"]] <- HiC.track(HiC.Midpoints = coaccplot.df, 
                                curr.locus.gr = curr.locus.gr, 
                                links.to.highlight = coaccplot.df$snp %in% pip.df$snp[pip.df$isAnnotate==T], curv = 1)

```

put it all together
```{r}
png('../manuscript_figures/HCN4_locus.png', width=1500, height=1200, res=150)
cowplot::plot_grid(plotlist = plot.list, align = 'v', ncol = 1, axis = 'b', rel_heights = c(1, 1, 1, 0.8, 0.8, 0.8, 1))
dev.off()
```


```{r}
plot.list[["pval.plot"]] <- plot.pval.track.y.axis(pip.df = pip.df, curr.locus.gr = curr.locus.gr)
plot.list[["pip.plot"]] <- plot.pip.track.w.y.axis(pip.df = pip.df, curr.locus.gr = curr.locus.gr)
png('../manuscript_figures/HCN4_y_locus.png', width=1500, height=1200, res=150)
cowplot::plot_grid(plotlist = plot.list, align = 'v', ncol = 1, axis = 'b', rel_heights = c(1, 1, 1, 0.8, 0.8, 0.8, 1))
dev.off()
```


