---
title: "Finemapping Results"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(liftOver)
library(ComplexHeatmap)
source('/project2/gca/aselewa/heart_atlas_project/R/analysis_utils.R')
```


### Finemapping results

I ran the script `R/run_finemapping.R` on each locus individually using `broadwl` for maximum parallelization. Below we load the results and concatenate. `finemappeR` was used to add prior probability to the GWAS summary statistics. 

```{r}
suppressMessages(library(tidyverse))
suppressMessages(library(finemappeR))

concat.fm.files <- function(file.names){
  
  stopifnot(length(file.names) > 1)
  rds.list <- list()
  for(f in file.names){
    rds.list[[f]] <- readRDS(f)
  }
  
  res <- Reduce(x = rds.list, f = rbind)
  
  res %>% as_tibble()
  
}


prior.files <- list.files(path = '../GWAS/finemapping/locus_files_finemapping', pattern = 'Finemapped_Prior*', full.names = T)
uniform.files <- list.files(path = '../GWAS/finemapping/locus_files_finemapping', pattern = 'Finemapped_Uniform*', full.names = T)


prior_res <- concat.fm.files(prior.files)
unif_res <- concat.fm.files(uniform.files)
```

```{r, eval=F}
saveRDS(prior_res, '../GWAS/finemapping/aFib_Finemapped.tble.rds')
```

```{r, eval=F}
prior_res <- readRDS('../GWAS/finemapping/aFib_Finemapped.tble.rds')
```


```{r}
finemap.CS.size <- prior_res %>% group_by(locus) %>% filter(CS == 1) %>% summarise(n = dplyr::n()) %>% .$n
```


```{r, fig.width=16, fig.height=2}
bks <- c(0, 1, 5, 10, 20, 50, 10000)
labs <- c("1", "2-5","6-10","11-20","21-50", "51+")
bin.count <- table(cut(finemap.CS.size, breaks = bks, labels = labs))
bin.count.df <- data.frame(count=as.numeric(bin.count), Size=factor(names(bin.count), levels = rev(labs)), group = "A", interval_size = c(2, 4, 4, 9, 29, 10))
bin.count.df$Size <- factor(labs, levels = labs)
```

```{r}
pdf('../manuscript_figures/Fig5B_unif_CS_distribution.pdf', width=6, height=4)
ggplot(bin.count.df, aes(x=Size, y = count)) + geom_bar(stat = 'identity',  fill='darkgreen', width=0.7) + ggClean() + xlab('Credible Set Size') + ylab('Credible Sets')
dev.off()
```


```{r}
ggplot(bin.count.df, aes(x = group, y=count, fill=Size)) + 
  geom_col() + 
  ggClean() + 
  xlab('Genetic Credible Set Size') + 
  ylab('Frequency') + 
  ggtitle('Number of causal variants / Number of loci') + 
  theme_void() +
  coord_flip() +
  geom_text(aes(label = paste0("\n", count)), color="white", position = position_stack(vjust = 0.5)) +
  geom_text(aes(label = paste0(Size, "\n")), position = position_stack(vjust = 0.5)) +
  theme(legend.position = 'none') +
  scale_fill_manual(values = c("#aaaaaa",brewer.pal(n = 6, name = "Greens")[-1]))
 
```

```{r, fig.width=16, fig.height=2}
bks <- c(0.01, 0.1, 0.5, 0.8, 0.9, 0.95, 0.99, 1)
labs <- c("1-10%","10-50%","50-80%","80-90%","90-95%","95-99%",">99%")
bin.count <- table(cut(prior_res$susie_pip, breaks = bks, labels = labs))
bin.count.df <- data.frame(count=as.numeric(bin.count), Size=factor(names(bin.count), levels = labs), group = "A", interval_size = c(10, 40, 30, 10, 6, 6, 4))
bin.count.df$Size <- factor(labs, levels = rev(labs))

```

```{r}
pdf('../manuscript_figures/Fig5B_uniform_pip_distribution.pdf', width=6, height=4)
ggplot(bin.count.df, aes(x=Size, y = count)) + geom_bar(stat = 'identity', width=0.7, fill='darkgreen') + ggClean() + xlab('PIP') + ylab('Number of SNPs') + coord_cartesian(ylim = c(0, 20))
dev.off()
```


```{r}
ggplot(bin.count.df, aes(x = group, y=interval_size, fill=Size)) + 
  geom_col() + 
  ggClean() + 
  xlab('Genetic Credible Set Size') + 
  ylab('Frequency') + 
  ggtitle('Posterior probability of being a causal variant') + 
  theme_void() +
  coord_flip() +
  geom_text(aes(label = paste0("\n", count)), color="white", position = position_stack(vjust = 0.5)) +
  geom_text(aes(label = paste0(Size, "\n")), position = position_stack(vjust = 0.5)) +
  theme(legend.position = 'none') +
  scale_fill_manual(values = c("#aaaaaa",brewer.pal(n = 7, name = "Greens")[-1]))


```


```{r}
sumstat.annot <- finemappeR::annotator(gwas = prior_res, annotations = list.files('../GWAS/annotations_for_finemapping_hg19/', '*.bed', full.names = T))
annots <- rep("Unannotated", nrow(sumstat.annot))
annots[sumstat.annot$all_eqtls_hg19.bed_d>0] <- "Other" #CHANGE BACK to eQTLs
annots[sumstat.annot$Coding_UCSC.bed_d>0] <- "Other"
annots[sumstat.annot$Conserved_LindbladToh.bed_d>0] <- "Other"
annots[sumstat.annot$CM_shared_peaks_hg19.bed_d>0] <- "CM Shared ATAC"
annots[sumstat.annot$non_CM_peaks_hg19.bed_d>0] <- "Non-CM ATAC"
annots[sumstat.annot$CM_specific_peaks_hg19.bed_d>0] <- "CM Specific ATAC"
```

```{r}
annot.snp.df <- data.frame(snp=prior_res$snp,
                           prior=prior_res$susie_pip, 
                          uniform=unif_res$susie_pip, 
                          annots=annots)
```

Here we visualize the effect of genomic annotations on the PIP of each SNP

```{r}
pdf('../manuscript_figures/Fig5B_PIP_vs_uniform.pdf', width=9, height=5)
ggplot(annot.snp.df, aes(x=uniform, y = prior, color = annots)) + 
  ggrastr::geom_point_rast() + 
    geom_abline(slope = 1, intercept = 0, col = 'black', linetype='dashed') +
  xlab('PIP (uniform prior)') + ylab('PIP (annotation prior)') + ggClean() +
  scale_color_manual(values = c("blue","firebrick","green","black","grey"))
dev.off()
```

### Linking causal SNPs to Genes

Filter finemapping results

```{r}
finemap.gr <- GenomicRanges::GRanges(seqnames = prior_res$chr, ranges = IRanges::IRanges(start = prior_res$pos, end = prior_res$pos))
finemap.gr$snp <- prior_res$snp
finemap.gr$pip <- prior_res$susie_pip
finemap.gr$chr <- prior_res$chr
finemap.gr$pos <- prior_res$pos
finemap.gr$locus <- prior_res$locus
seqlevelsStyle(finemap.gr) <- "UCSC"
finemap.gr <- finemap.gr[finemap.gr$pip > 1e-5, ]
```

load annotations for linking SNPs to genes (run `clean_genomic.annotations.R` to produce)

```{r}
genomic.annots <- readRDS('../genomic_annotations/hg19_gtf_genomic_annots.gr.rds')
gene.annots <- genomic.annots$genes
genomic.annots$genes <- NULL
genomic.annots$introns <- NULL
splice.juncs <- genomic.annots$splice_junctions
genomic.annots$splice_junctions <- NULL
```

Download Hi-C data from https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-6014/ and convert the dataframe into a GRanges object.

```{r}
genomic.annots$HiC <- readRDS('../HiC/iPSC_CM_pcHiC_protein_Hg19.gr.rds')
```

Add co-accessibility by running the script `R/prepare_Coaccessibility.R` and load it here

```{r}
coaccess <- readRDS('../ArchR/ArchR_heart_latest_noAtrium/CoAccessibility/Coacc_ENHANCERS_AllCellTypes_overlapCutoff50_k200_corr_cut_-1_maxDist_1Mb_hg19.gr.rds')
coaccess <- coaccess[coaccess$correlation>0.2,]
genomic.annots$coacc <- coaccess
```

Overlap finemapped SNPs with each of our annotations

```{r}
# Heirarchy level 1: HiC, CoAccess, exon, promoter, UTR
snps.overlap.annots <- GRangesList(lapply(genomic.annots, function(x) { plyranges::join_overlap_inner(x = x, y = finemap.gr) }))
snps.in <- unique(unlist(snps.overlap.annots)$snp)
snps.unassigned.gr <- finemap.gr[!(finemap.gr$snp %in% snps.in),]

# Heirarchy level 2: splice junction
snps.overlap.annots$splice_junction <- plyranges::join_overlap_inner(x = splice.juncs, y = snps.unassigned.gr)
snps.in <- unique(unlist(snps.overlap.annots)$snp)
snps.unassigned.gr <- finemap.gr[!(finemap.gr$snp %in% snps.in),]

# Heirarchy level 3: intron/intragenic
res <- plyranges::join_overlap_inner(x = genomic.annots$promoters, 
                                     y = snps.unassigned.gr, maxgap = 1e6) # get all promoters within 1MB

res$distance <- abs(res$pos - end(res))
res$weight <- exp(-res$distance / 1e5)

snps.overlap.annots$intronic <- res
```

```{r}
categories <- names(snps.overlap.annots)
mat.list <- list()

for(c in categories){
  curr <- snps.overlap.annots[[c]]@elementMetadata %>% as_tibble()
  if(c == "intronic"){
    curr <- curr[,c("snp","chr", "pos","locus","pip","gene_name", "weight")] %>% mutate(category = "distance")
  }
  else if(c == "splice_junction"){
    curr <- curr[,c("snp","chr", "pos","locus","pip","gene_name", "weight")] %>% mutate(category = "splice_junction", weight=0.5)
  }
  else{
    curr <- curr[,c("snp","chr", "pos","locus","pip","gene_name", "weight")] %>% mutate(category = c, weight = 1*(pip>0))
  }
  mat.list[[c]] <- curr
} 

final.mat <- Reduce(rbind, mat.list) %>% distinct(gene_name, snp, pip, .keep_all = TRUE)
```

```{r}
final.mat <- final.mat %>% group_by(snp) %>% mutate(frac_pip = weight/sum(weight))
final.mat <- final.mat %>% group_by(gene_name) %>% mutate(gene_pip = sum(pip * frac_pip)) %>% arrange(gene_name)

snp.annots <- data.frame(snp=prior_res$snp, annots=annots)
final.mat <- dplyr::left_join(x = final.mat, y = snp.annots, by = "snp")
final.mat <- final.mat[!is.na(final.mat$gene_name),]
```

```{r}
saveRDS(final.mat, '../GWAS/finemapping/aFib_Finemapped_GeneMapped_06152021.tble.rds')
```

### prepare tables for experimental verification

```{r}
gene.view.df <- final.mat %>% 
    dplyr::select(locus, gene_name, gene_pip, pip) %>% 
    group_by(locus, gene_name) %>% 
    summarise(gene_pip = round(gene_pip[1], 3), n_snps = sum(pip > 0.01)) %>% 
    filter(gene_pip > 0.1) %>%
    readr::write_csv('../GWAS/aFib_Finemapped_GenePIP_0.1_06152021.csv')
    

high.conf.snp.df <- final.mat %>% 
    dplyr::filter(pip > 0.2) %>% 
    group_by(snp) %>% 
    arrange(-gene_pip) %>% 
    dplyr::slice(1) 
```

```{r}
snp.gene <- high.conf.snp.df %>% dplyr::select(snp, pos, gene_name)
gene.locs.df <- gene.annots %>% as_tibble()
gene.locs.df$TSS <- ifelse(gene.locs.df$strand=="+", gene.locs.df$start, gene.locs.df$end)

snp.gene.dist <- snp.gene %>% left_join(., gene.locs.df, on = 'gene_name') %>%
    mutate(dist = abs(TSS - pos)) %>%
    dplyr::select(snp, gene_name, dist)

high.conf.snp.df <- inner_join(high.conf.snp.df, snp.gene.dist, on = c('snp','gene_name'))
```

I downloaded ALL H3K27ac bed files from Encode for Aorta, Heart LV/RV/RA, and cardiac muscle. I concatenated them into one bed file. They were in Hg38 so I converted them to Hg19 with liftOver. 

```{r}
seqlevelsStyle(finemap.gr) <- "NCBI"
files <- list.files(path = '../ENCODE/H3k27ac_gwas_hg19/hg19_mapped', pattern = '*.bed', full.names = T)
files <- c(files, '../ENCODE/H3k27ac_gwas_hg19/FetalHeart_E083-DNase_hg19_cleaned_narrowPeak.bed.gz', '../ENCODE/FGT_ChIP_lifted_from_mm10.bed')
z <- lapply(files, function(x){rtracklayer::import(x)})
annotated.snps <- lapply(files, function(x){get_elements_overlap_snps(snp.gr = finemap.gr, annotations = x)})
```

```{r}
for(i in 1:length(annotated.snps)){
    high.conf.snp.df <- left_join(high.conf.snp.df, annotated.snps[[i]], on = 'snp')
}
```

SNP-level table at >20% PIP

```{r}
high.conf.snp.df %>% 
    dplyr::select(-weight, -frac_pip, -locus) %>% 
    dplyr::rename(SNP = snp, `b37 bp` = pos, PIP = pip, `Gene Linked` = gene_name, 
                  `Gene PIP`=gene_pip, `Link Method`=category, `Distance to Gene`=dist, `Chromatin status`=annots) %>% 
    mutate(PIP = round(PIP, 3), `Gene PIP` = round(`Gene PIP`, 3)) %>% 
    arrange(chr, `b37 bp`) %>% 
    readr::write_csv('../GWAS/aFib_Finemapped_minPIP_0.2_06152021.csv')
```

```{r}

nielsen_snp_genes <- suppressMessages(readr::read_csv('../GWAS/Nielsen_snps_prior_genes.csv', col_names = T)) %>% dplyr::rename(snp = rsID)
snp.pos <- as.integer(sub("chr.*:","",nielsen_snp_genes$`Position (hg19)`))
top.snp.gr <- GRanges(seqnames = sub(":.*","",nielsen_snp_genes$`Position (hg19)`), ranges = IRanges(start = snp.pos, end = snp.pos)) + 1e6
top.snp.gr$snp <- nielsen_snp_genes$snp
top.snp.gr$nielsen_locus <- nielsen_snp_genes$`Locus No.`
top.snp.gr$genes_prioritized <- nielsen_snp_genes$`Prioritized genes`

data('Euro_LD_Chunks', package='finemappeR')
ldblocks.gr <- GRanges(seqnames = paste0("chr",LD_Blocks$X1), ranges = IRanges(start = LD_Blocks$X2, end = LD_Blocks$X3), locus=LD_Blocks$X4)

locus.convert <- plyranges::join_overlap_inner(top.snp.gr, ldblocks.gr) %>% as_tibble() %>% dplyr::select(nielsen_locus, locus, genes_prioritized)

```

Locus/gene view

```{r}
gene.cumsum.df <- gene.view.df %>% 
    group_by(locus) %>% 
    arrange(-gene_pip) %>% mutate(gene_pip_csum = cumsum(gene_pip)) %>% 
    dplyr::slice(1:which(gene_pip_csum >= 0.8)[1]) 

gene.cs.df <- gene.cumsum.df %>% 
    group_by(locus) %>% 
    summarise(gene_cs = paste0(gene_name, collapse=','), top_gene = gene_name[1], top_gene_pip = gene_pip[1])

gene_view_df <- left_join(gene.cumsum.df, locus.convert, by = 'locus') %>% 
    ungroup() %>%
    dplyr::rename(Locus = nielsen_locus, `Gene Name` = gene_name, `Gene PIP` = gene_pip, `N SNPs Linked` = n_snps) %>%
    dplyr::select(Locus, `Gene Name`, `Gene PIP`, `N SNPs Linked`) %>%
    write_csv('../GWAS/aFib_Finemapped_GeneView.csv')

locus_view_df <- left_join(gene.cs.df, locus.convert, by = 'locus') %>% 
    group_by(nielsen_locus) %>% 
    summarise(`Nielsen Gene` = paste0(unique(genes_prioritized), collapse=','),
              `80% Gene Credible Set` =  paste0(unique(gene_cs), collapse=','),
              `Top Genes` = paste0(unique(top_gene), collapse=','),
              `Top Gene PIP` = paste0(unique(top_gene_pip), collapse=',')) %>%
    dplyr::rename(Locus = nielsen_locus) %>%
    write_csv('../GWAS/aFib_Finemapped_LocusView.csv')
```

```{r}
final.mat.high.conf <- left_join(final.mat.high.conf, nielsen_snp_genes, by = 'snp') 

final.mat.high.conf[is.na(final.mat.high.conf)] <- ""
final.mat.high.conf$is_prioritized <- ifelse(yes = "Yes", no = "No", test = (final.mat.high.conf$gene_name  == final.mat.high.conf$`Prioritized genes`))
```

```{r}
final.mat.high.conf %>% dplyr::select(-weight, -frac_pip) %>% arrange(chr, pos, -pip) %>% write_csv('../GWAS/aFib_Finemapped_MinPip0.2_03232021.csv')
```

### Gene PIP distribution

```{r}
finemap.res <- readRDS('../GWAS/finemapping/aFib_Finemapped_GeneMapped.tble.rds')
finemap.res <- finemap.res %>% group_by(gene_name) %>% slice(1)
bks <- c(0.1, 0.5, 0.8, 0.9, 0.95, 0.99, 100)
labs <- c("10-50%","50-80%","80-90%","90-95%","95-99%",">99%")
bin.count <- table(cut(finemap.res$gene_pip, breaks = bks, labels = labs))
bin.count.df <- data.frame(count=as.numeric(bin.count), Size=factor(names(bin.count), levels = labs))
bin.count.df$Size <- factor(labs, levels = rev(labs))
```

```{r}
pdf('../manuscript_figures/Fig5G_gene_pip_distribution.pdf', width=6, height=4)
ggplot(bin.count.df, aes(x=Size, y = count)) + geom_bar(stat = 'identity',  fill='lightgreen', width=0.7) + ggClean() + xlab('Gene PIP') + ylab('Number of Genes') + coord_cartesian(ylim = c(0,50))
dev.off()
```


### Gene CS size Distribution

code below is wrong, fix edge case when 1 SNP has a PIP higher than the CS limit
```{r}
cs.thresh <- 0.9
finemap.res <- readRDS('../GWAS/finemapping/aFib_Finemapped_GeneMapped.tble.rds')
gene.cs <- finemap.res %>%
  dplyr::select(c(gene_name,gene_pip,locus)) %>%
  distinct() %>%
  group_by(locus) %>%
  arrange(desc(gene_pip)) %>%
  mutate(gene_pip_cumsum = cumsum(gene_pip)) %>%
  filter(gene_pip_cumsum <= 0.9)
gene.cs.size <- gene.cs %>% group_by(locus) %>% summarise(n = dplyr::n()) %>% .$n

```

```{r, fig.width=16, fig.height=2}
bks <- c(0, 1, 5, 10, 15, 10000)
labs <- c("1", "2-5","6-10","11-15","16+")
bin.count <- table(cut(gene.cs.size, breaks = bks, labels = labs))
bin.count.df <- data.frame(count=as.numeric(bin.count), Size=factor(names(bin.count), levels = rev(labs)))
bin.count.df$Size <- factor(labs, levels = labs)

pdf('../manuscript_figures/Fig5G_gene_cs_distribution.pdf', width=6, height=4)
ggplot(bin.count.df, aes(x=Size, y = count)) + geom_bar(stat = 'identity',  fill='lightgreen', width=0.7) + ggClean() + xlab('Gene Credible Set Size') + ylab('Gene Credible Sets')
dev.off()
```

```{r}
finemap.res <- readRDS('../GWAS/finemapping/aFib_Finemapped_GeneMapped_06152021.tble.rds')
celltype_ideal_order <- c("Cardiomyocyte","Smooth Muscle","Pericyte","Endothelial","Fibroblast","Neuronal", "Lymphoid","Myeloid")
srna <- readRDS('../seurat/Heart_RNA_Processed_Combined_NoAtrium.rds')
srna.exp <- Seurat::AverageExpression(srna)
srna.exp.mat <- as.matrix(log2(srna.exp$RNA + 1))
srna.exp.mat <- srna.exp.mat[,celltype_ideal_order]
srna.exp.mat.z <- sweep(srna.exp.mat-rowMeans(srna.exp.mat), MARGIN = 1, STATS = matrixStats::rowSds(srna.exp.mat), FUN = '/')

finemap.res <- finemap.res[finemap.res$gene_pip > 0.5,]
finemap.res <- finemap.res %>% group_by(gene_name) %>% dplyr::slice(1)
finemap.genes.exp <- srna.exp.mat[intersect(finemap.res$gene_name, rownames(srna.exp.mat)),celltype_ideal_order]
finemap.genes.exp.z <- srna.exp.mat.z[intersect(finemap.res$gene_name, rownames(srna.exp.mat)),celltype_ideal_order]
```

```{r}
pdf('../manuscript_figures/figure5/Finemap_Genes_Exp.pdf', width = 9, height=3)
ComplexHeatmap::Heatmap(t(finemap.genes.exp.z), cluster_rows  = T, show_column_dend = F,
                        col = circlize::colorRamp2(c(-2, 0, 2), c("lightblue", "white","firebrick")))
dev.off()
```

```{r}
finemap.genes.exp[finemap.genes.exp > 4] <- 4
finemap.genes.exp.long <- finemap.genes.exp %>% as_tibble() %>% pivot_longer(cols = everything())
finemap.genes.exp.long$name <- factor(finemap.genes.exp.long$name, levels = celltype_ideal_order)
palette <- readRDS('palette.rds')
ggplot(finemap.genes.exp.long, aes(x=name, y = value, fill=name)) + geom_violin() + geom_boxplot(width=0.1) + ggClean(rotate_axis = T) + ylab('log2 transcripts per 10k') + xlab('Cell Type') + 
    scale_fill_manual(values = palette) + LegendOff()
```

```{r}
fetal.heart.exp <- as.data.frame(read_csv('../Atria-RNA-seq-fetal-counts.csv'))
fetal.heart.exp$Gene_ENSEMBL <- NULL
fetal.heart.exp <- fetal.heart.exp[!duplicated(fetal.heart.exp$Gene_symbol),]
fetal.heart.exp <- fetal.heart.exp[!is.na(fetal.heart.exp$Gene_symbol),]

rownames(fetal.heart.exp) <- fetal.heart.exp$Gene_symbol
fetal.heart.exp$Gene_symbol <- NULL
fetal.heart.exp <- fetal.heart.exp[!is.na(rowSums(fetal.heart.exp)),]
fetal.heart.norm <- t(log2((t(fetal.heart.exp)/colSums(fetal.heart.exp))*10^6 + 1))
```

```{r}
fetal.heart.norm.gwas <- fetal.heart.norm[match(rownames(finemap.genes.exp), rownames(fetal.heart.norm)),]

plot(fetal.heart.norm.gwas[,'LA-fetal-22w'], finemap.genes.exp[,'Cardiomyocyte'], xlab='log2 CPM (fetal 14w)', ylab='log2 TP10k (CM RNA)', pch=16)
```

```{r}
fetal.exp <- as.data.frame(fetal.heart.norm.gwas[,c('LA-fetal-14w','LA-fetal-22w','RA-fetal-14w','RA-fetal-22w')])
fetal.exp$gene_name <- rownames(fetal.exp)

fm.gene.df <- as.data.frame(finemap.genes.exp[,'Cardiomyocyte',drop=F])
fm.gene.df$gene_name <- rownames(fm.gene.df)

fm.gene.z.df <- as.data.frame(finemap.genes.exp.z[,'Cardiomyocyte',drop=F])
fm.gene.z.df$gene_name <- rownames(fm.gene.z.df)
colnames(fm.gene.z.df) <- c("Cardiomyocyte_Zscore","gene_name")

pip.df <- finemap.res[,c("gene_name", "gene_pip")]

left_join(fm.gene.df, pip.df, on='gene_name') %>% left_join(., fm.gene.z.df) %>% left_join(., fetal.exp) %>% dplyr::select(gene_name, gene_pip, Cardiomyocyte, Cardiomyocyte_Zscore, `LA-fetal-14w`,`LA-fetal-22w`,`RA-fetal-14w`,`RA-fetal-22w`) %>% arrange(Cardiomyocyte) %>% write_tsv('../GWAS/finemapping/aFig_genes_pip_50_fetal_comparison.tsv')
```


```{r}
final.mat <- readr::read_tsv('../GWAS/aFib_Finemapped_minPIP_0.2_04202021_CardiacTF_FGT.csv')
final.mat <- final.mat[final.mat$PIP > 0.5,]
```

```{r}
prop.chromatin <- as.numeric(table(final.mat$`Chromatin status`))
prop.chromatin <- prop.chromatin[-4]
prop.h3k27ac <- colSums(is.na(final.mat[,startsWith(prefix = 'Hg19', x = colnames(final.mat))]))
prop.h3k27ac <- prop.h3k27ac[c(3,5,4)]
prop.FGT <- sum(!is.na(final.mat$FGT_ChIP_lifted_from_mm10))
prop.cardiacTF <- sum(!is.na(final.mat$tf.info))
```

```{r}
snp.activity.df <- data.frame(props = c(prop.chromatin, prop.h3k27ac, prop.cardiacTF), names  = c("CM Shared","CM Specific", "Non-CM","Heart LV","Heart RV", "Heart RA","Cardiac Motifs"), type = c(rep("ATAC",3), rep("H3K27ac",3),"Motif Scan"))
snp.activity.df$names <- factor(snp.activity.df$names, 
                                levels = snp.activity.df$names[c(2,1,3,4,5,6,7,8)])
snp.activity.df$type <- factor(snp.activity.df$type, levels = c("ATAC","H3K27ac","Motif Scan"))
```


```{r}
pdf('../manuscript_figures/figure5/Fig5E_top_pip_snp_overlap.pdf', width=8, height=6)
ggplot(snp.activity.df, aes(x = names, y = props, fill = type)) + 
    geom_bar(stat='identity') + 
    ylab('Number of SNPs (of 54)') + 
    xlab('') + 
    ggClean(rotate_axis = T) +
    scale_fill_manual(values = c("lightblue","navy","purple"))
dev.off()
```

