---
title: "Heart Atlas - Combined"
author: "Alan Selewa"
data: "10/30/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Introduction

10X snRNA-seq and snATAC-seq was ran on 6 regions of the heart from 3 individuals. Here we explore the clustering results from combining all donors and heart chambers for both modalities. Generally the data were good quality. We removed cells with small libraries (< 1000 UMI in RNA, < 3000 unique fragments in ATAC). We also removed doublets in ATAC-seq using ArchR's doublet removal scheme. For RNA-seq, we removed doublets using the DoubletFinder tool.

### Clustering

The dimensionality reduction and clustering was done outside of this notebook and the resulting objects were serialized/saved. Details as to how the data were processed are in `R/`.

```{r}
suppressMessages(library(Seurat))
suppressMessages(library(ArchR))
suppressMessages(library(viridis))
suppressMessages(source('../R/analysis_utils.R'))

srna <- readRDS('../seurat/Heart_RNA_Processed_Combined.rds')
satac <- suppressMessages(loadArchRProject('../ArchR/ArchR_heart/', showLogo = F))
```

We can visualize the six different regions:

```{r, fig.width=10, fig.height=5}
palette <- setNames(brewer.pal(6, "Set2"), unique(srna$region))
p1 <- DimPlot(srna, label = F, group.by='region', cols = palette) + ggClean() + ggtitle('scRNA-seq')
p2 <- custom_archr_umap(archr_project = satac, group.by = 'regions', palette = palette, legend = F, label = F, pt.size = 0.05) + ggtitle('scATAC-seq')
p1 + p2
```


We can visualize the clustering results with the inferred cell types. I used the following markers to define cell types: 

```{r, echo = F}
markers.df <- data.frame(marker = c('TTN/MYBPC3',"MYH7","NPPA",'RGS5/ABCC9','MYH11/TAGLN','DCN/PDGFRA','PECAM1/VWF','PLP1','CDA8/LCK','CD14/FOLR2'),
                         CellType = c('Cardiomyocytes','Vent. CM','Atrial CM','Pericyte','Smooth Muscle','Fibroblast','Endothelial','Neuronal','Lymphoid','Myeloid'))
knitr::kable(markers.df, caption = 'Heart cell markers')
```


```{r, warning=FALSE, fig.width=14, fig.height=8}
palette <- readRDS('palette.rds')
p1 <- DimPlot(srna, label = T, cols = palette) + ggClean() + NoLegend() + ggtitle('scRNA-seq')
p2 <- custom_archr_umap(archr_project = satac, group.by = 'CellTypes', palette = palette, legend = F, label = T) + ggtitle('scATAC-seq')
p1 + p2
```


We recover 9 cell types in scRNA-seq and 7 in scATAC-seq. How do they compare in proportion?

As expected, cardiomyocytes are the dominant cell type in both assays.

```{r}
freq <- table(Idents(srna))/length(Idents(srna))
p1 <- make_freq_plot(freq, palette = palette) + LegendOff() + ggtitle('scRNA-seq')

freq <- table(satac$CellTypes)/length(satac$CellTypes)
p2 <- make_freq_plot(freq, palette = palette) + LegendOff() + ggtitle('scATAC-seq')

p1 + p2
```

### CRE discovery

Here we run macs2 on ATAC-seq aggregates of each cell-type to discover cell-type specific peaks. Below shows the number of peaks called in total and for each cell type. As expected, most peaks are not protein coding.

```{r}
peak.info <- getPeakSet(satac)
peak.info.per <- peak.info@metadata$PeakCallSummary

pal <- c("Distal" = "#60BA64", "Exonic" = "#73C6FF", "Intronic" = "#620FA3", "Promoter" = "#FFC554")
ggplot(peak.info.per, aes(x=Group, y=Freq, fill=Var1)) + geom_bar(stat='identity') + labs(fill='') +
  scale_fill_manual(values = pal) + ggClean(rotate_axis = T) + ylab("Number of Peaks \n (10^3)")
```

Based on the union set, we create a single-cell binary peak accessibility matrix and find the peaks that are differentially accessible. 

### Differential accessibility analysis

#### Approach

I used two approaches for differential accessibility: 

* one vs all single-cell differential accessibility (wilcoxon rank sum test)
* one vs all pseudo-bulk differential accessibility in edgeR (as was done in Hocker et al.)

```{r}
archrPeaks <- readRDS('../ArchR/ArchR_heart/PeakCalls/DA_markerPeaks.rds')
edgePeaks <- readRDS('../ArchR/ArchR_heart/PeakCalls/edgeR_DA_peaks.gr.rds')
```

```{r}
edgePeaks <- sapply(edgePeaks, function(x){sort(sortSeqlevels(x))})
fdr.df <- as.data.frame(sapply(edgePeaks, function(x){x$FDR}))
fc.df <- as.data.frame(sapply(edgePeaks, function(x){x$Log2FC}))
fdr.df <- fdr.df[,archrPeaks@colData@rownames]
fc.df <- fc.df[,archrPeaks@colData@rownames]

edgeRpeaks <- archrPeaks
edgeRpeaks@assays@data$Log2FC <- fc.df
edgeRpeaks@assays@data$FDR <- fdr.df
```

#### Global comparison

Both methods generally detect the same peaks differentially accessible peaks. Below is a plot of the log fold change from edgeR and from ArchR. 

```{r}
arch.log2fc <- assays(archrPeaks)[['Log2FC']]$`Vent. CM`
arch.fdr <- assays(archrPeaks)[['FDR']]$`Vent. CM`
edge.log2fc <- assays(edgeRpeaks)[['Log2FC']]$`Vent. CM`
edge.fdr <- assays(edgeRpeaks)[['FDR']]$`Vent. CM`

logfc.df <- data.frame(ArchR_logfc=arch.log2fc, edgeR_logfc=edge.log2fc)
get_density <- function(x, y, ...) {
  dens <- MASS::kde2d(x, y, ...)
  ix <- findInterval(x, dens$x)
  iy <- findInterval(y, dens$y)
  ii <- cbind(ix, iy)
  return(dens$z[ii])
}
logfc.df$density <- get_density(logfc.df$ArchR_logfc, logfc.df$edgeR_logfc)
ggplot(logfc.df, aes(x=ArchR_logfc, y=edgeR_logfc, color=density)) + geom_point() + xlab('edgeR log2FC') + ylab('ArchR log2FC') + geom_abline(slope = 1, color="red", linetype="dashed") + ggClean() + scale_color_viridis() + ggtitle('Vent. CM ') + ggClean()
```

#### Reproducibility across donors

Here I test which method yields better reproducibility across donors. Briefly, I obtained differentially accessible peaks from each donor (n = 3) using the ArchR and edgeR methods. Below I summarize how the Vent CM peaks are shared across donors with the two methods. 

```{r}
projs <- list.files(path = '../ArchR', pattern = 'ArchR_project_*', full.names = T)
archr.list <- list()
edger.list <- list()
for( i in 1:length(projs)){
  curr.edger.peaks <- readRDS(paste0(projs[i], '/PeakCalls/edgeR_DA_peaks.gr.rds'))$`Vent. CM`
  curr.edger.peaks <- curr.edger.peaks[curr.edger.peaks$FDR <= 0.01 & curr.edger.peaks$Log2FC >= 1]
  curr.archr.peaks <- readRDS(paste0(projs[i], '/PeakCalls/DA_MARKERS_FDRP_1_log2FC_1.rds'))$`Vent. CM`
  
  curr.edger.peaks$donor <- i
  curr.archr.peaks$donor <- i
  edger.list[[i]] <- curr.edger.peaks
  archr.list[[i]] <- curr.archr.peaks
}
```

```{r}
edger.share <- getSharingMat(edger.list)
archr.share <- getSharingMat(archr.list)
```

As we can see below, the sharing is generally higher using ArchRs method, which is somewhat expected given that ArchR has a built in method for finding reproducible peaks.

```{r results = 'hide'}
a <- pheatmap::pheatmap(edger.share, cluster_cols = F, cluster_rows = F, main = 'CM Peak Sharing (edgeR)', silent = T)[[4]]
b <- pheatmap::pheatmap(archr.share, cluster_cols = F, cluster_rows = F, main = 'CM Peak Sharing (ArchR)', silent = T)[[4]]
```

```{r}
gridExtra::grid.arrange(gridExtra::arrangeGrob(grobs= list(a=a,b=b),ncol=2))
```


Conclusion: ArchR differentially accessible peaks are more reproducible across donors than edgeR peaks. 

### Motif enrichment

Next we compare these two peak sets in terms of motif enrichment. Below we plot the enrichment of motifs in Vent. CM peaks from both methods using *aggregate donor* data for each cell-type. 

```{r}
archr_motifUp <- suppressMessages(peakAnnoEnrichment(seMarker = archrPeaks, ArchRProj = satac, peakAnnotation = "Motif", cutOff = "FDR <= 0.01 & Log2FC >= 0.5"))
edger_motifup <- suppressMessages(peakAnnoEnrichment(seMarker = edgeRpeaks, ArchRProj = satac, peakAnnotation = "Motif", cutOff = "FDR <= 0.01 & Log2FC >= 0.5"))
```

```{r}
motifScatterPlot <- function(motifUp1, motifUp2){
 df1 <- data.frame(TF1 = rownames(motifUp1), mlog10Padj1 = assay(motifUp1)[,7])
 df2 <- data.frame(TF2 = rownames(motifUp2), mlog10Padj2 = assay(motifUp2)[,7])
 df <- cbind(df1, df2)
 
 ggplot(df, aes(mlog10Padj1, mlog10Padj2)) + 
   geom_point(size = 1) +
   ggrepel::geom_label_repel(
     data = df[df$mlog10Padj1 > 100 | df$mlog10Padj2 > 100, ], aes(x = mlog10Padj1, y = mlog10Padj2, label = TF1), 
     size = 3,
     nudge_x = 2,
     color = "black"
   ) + ggClean() + 
   ylab("-log10(P-adj) (edgeR)") + 
   xlab("-log10(P-adj) (ArchR)") +
   LegendOff()
}
```

```{r}
motifScatterPlot(archr_motifUp, edger_motifup) + ggtitle('Motif Enrichment (Vent CM)')
```

### Exploration: GTEx enrichment

GTEx has eQTL data for n = 495 individuals for heart tissue (left ventricle.) We hypothesize that the cardiomyocyte (CM) regulatory elements are enriched with heart eQTLs. We use Torus to perform this enrichment analysis. We plot the log2 enrichment result below. 

```{r}
enrich.df <- read.delim('../eQTL_enrich/results/torus_celltypes_combined.enrichment', header=F, sep="")

ggplot(enrich.df, aes(x=V2, y=V1)) + geom_point() + geom_errorbar(aes(xmin=V3, xmax=V4), colour="black", width=.1) + ggClean() +
  xlab('log2 Enrichment') + ylab('Cell Type') + geom_vline(xintercept = 0, col='red', linetype = "longdash") + ggtitle('Enrichment of ArchR DA Peaks')
```

```{r}
enrich.df <- read.delim('../eQTL_enrich/results/DA_peaks_edgeR.enrichment', header=F, sep="")

ggplot(enrich.df, aes(x=V2, y=V1)) + geom_point() + geom_errorbar(aes(xmin=V3, xmax=V4), colour="black", width=.1) + ggClean() +
  xlab('log2 Enrichment') + ylab('Cell Type') + geom_vline(xintercept = 0, col='red', linetype = "longdash") + ggtitle('Enrichment of edgeR DA Peaks')
```





