---
title: "Heart Atlas - Combined"
author: "Alan Selewa"
data: "10/30/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Introduction

10X snRNA-seq and snATAC-seq was ran on 6 regions of the heart from 3 individuals. Here we explore the clustering results from combining all donors and heart chambers for both modalities. Generally the data were good quality. We removed cells with small libraries (< 1000 UMI in RNA, < 3000 unique fragments in ATAC). We also removed doublets in ATAC-seq using ArchR's doublet removal scheme. For RNA-seq, we removed doublets using the DoubletFinder tool.

### Clustering

The dimensionality reduction and clustering was done outside of this notebook and the resulting objects were serialized/saved. Details as to how the data were processed are in `R/`.

```{r}
suppressMessages(library(Seurat))
suppressMessages(library(ArchR))
suppressMessages(source('../R/analysis_utils.R'))

srna <- readRDS('../seurat/Heart_RNA_Processed_Combined.rds')
satac <- suppressMessages(loadArchRProject('../ArchR/ArchR_heart/', showLogo = F))
```

We can visualize the six different regions:

```{r, fig.width=10, fig.height=5}
palette <- setNames(brewer.pal(6, "Set2"), unique(srna$region))
p1 <- DimPlot(srna, label = F, group.by='region', cols = palette) + ggClean() + ggtitle('scRNA-seq')
p2 <- custom_archr_umap(archr_project = satac, group.by = 'regions', palette = palette, legend = F, label = F, pt.size = 0.05) + ggtitle('scATAC-seq')
p1 + p2
```


We can visualize the clustering results with the inferred cell types. I used the following markers to define cell types: 

```{r, echo = F}
markers.df <- data.frame(marker = c('TTN/MYBPC3',"MYH7","NPPA",'RGS5/ABCC9','MYH11/TAGLN','DCN/PDGFRA','PECAM1/VWF','PLP1','CDA8/LCK','CD14/FOLR2'),
                         CellType = c('Cardiomyocytes','Vent. CM','Atrial CM','Pericyte','Smooth Muscle','Fibroblast','Endothelial','Neuronal','Lymphoid','Myeloid'))
knitr::kable(markers.df, caption = 'Heart cell markers')
```


```{r, warning=FALSE, fig.width=14, fig.height=8}
palette <- readRDS('palette.rds')
p1 <- DimPlot(srna, label = T, cols = palette) + ggClean() + NoLegend() + ggtitle('scRNA-seq')
p2 <- custom_archr_umap(archr_project = satac, group.by = 'CellTypes', palette = palette, legend = F, label = T) + ggtitle('scATAC-seq')
p1 + p2
```


We recover 9 cell types in scRNA-seq and 7 in scATAC-seq. How do they compare in proportion?

As expected, cardiomyocytes are the dominant cell type in both assays.

```{r}
freq <- table(Idents(srna))/length(Idents(srna))
p1 <- make_freq_plot(freq, palette = palette) + LegendOff() + ggtitle('scRNA-seq')

freq <- table(satac$CellTypes)/length(satac$CellTypes)
p2 <- make_freq_plot(freq, palette = palette) + LegendOff() + ggtitle('scATAC-seq')

p1 + p2
```

### CRE discovery

Here we run macs2 on ATAC-seq aggregates of each cell-type to discover cell-type specific peaks. Below shows the number of peaks called in total and for each cell type. As expected, most peaks are not protein coding.

```{r}
peak.info <- getPeakSet(satac)
peak.info.per <- peak.info@metadata$PeakCallSummary

pal <- c("Distal" = "#60BA64", "Exonic" = "#73C6FF", "Intronic" = "#620FA3", "Promoter" = "#FFC554")
ggplot(peak.info.per, aes(x=Group, y=Freq, fill=Var1)) + geom_bar(stat='identity') + labs(fill='') +
  scale_fill_manual(values = pal) + ggClean(rotate_axis = T) + ylab("Number of Peaks \n (10^3)")
```

Based on the union set, we create a single-cell binary peak accessibility matrix and find the peaks that are differentially accessible. Below we plot the relative accessiblility across cells for differentially accessible peaks. We only plot peaks with FDR < 10% and log2 fold-change > 0.5. Based on this filtering, we have discovered 245,901 regulatory elements at 10% FDR. Specially, 88,610 peaks were upregulated in ventricular cardiomyocytes. 

#### Differential accessibility analysis

```{r}
markersPeaks <- readRDS('../ArchR/ArchR_heart/PeakCalls/DA_markerPeaks.rds')
suppressMessages(plotMarkerHeatmap(seMarker = markersPeaks, cutOff = "FDR <= 0.01 & Log2FC >= 1", transpose = TRUE))
```

```{r}
peak.mat <- suppressMessages(getMatrixFromProject(satac, useMatrix = "PeakMatrix"))
peak.mat <- peak.mat@assays@data$PeakMatrix
all.peaks.gr <- getPeakSet(satac)
all.peaks <- GRToString(all.peaks.gr)
row.names(peak.mat) <- all.peaks
```


### Exploration: GTEx enrichment

GTEx has eQTL data for n = 495 individuals for heart tissue (left ventricle.) We hypothesize that the cardiomyocyte (CM) regulatory elements are enriched with heart eQTLs. We use Torus to perform this enrichment analysis. We plot the log2 enrichment result below. 

```{r}
enrich.df <- read.delim('../eQTL_enrich/results/torus_celltypes_combined.enrichment', header=F, sep="")

ggplot(enrich.df, aes(x=V2, y=V1)) + geom_point() + geom_errorbar(aes(xmin=V3, xmax=V4), colour="black", width=.1) + ggClean() +
  xlab('log2 Enrichment') + ylab('Cell Type') + geom_vline(xintercept = 0, col='red', linetype = "longdash")
```

### Exploration: signal explained in finemapped heart eQTLs

As expected, ventricular cardiomyocyte peaks are enriched with heart eQTLs, with endothelial peaks as second most enriched. 

```{r}
finemap.pips <- read.delim('../eQTL_enrich/gtex_finemapping/GTEx_v8_finemapping_DAPG/SNP_PIP_df.txt', sep='\t',header=T)
chr <- sub(pattern='_.*', replacement = "", x = finemap.pips$SNP)
start <- sub(pattern='chr[0-9]+_', replacement = "", x = finemap.pips$SNP)
start <- as.numeric(sub(pattern='_.*', replacement = "", x = start))
finemap.pip.gr <- GenomicRanges::GRanges(seqnames = chr, IRanges::IRanges(start = start, end = start), pip = finemap.pips$PIP)
```

```{r}
gr.list <- readRDS('../ArchR/ArchR_heart/PeakCalls/DA_MARKERS_FDRP_1_log2FC_1.rds')
```

```{r}
percent.snps <- rep(0, length(gr.list))
percent.pip <- rep(0, length(gr.list))
enrich <- rep(0, length(gr.list))
for(i in 1:length(gr.list)){
  curr.peaks.gr <- gr.list[[i]]
  x <- IRanges::subsetByOverlaps(finemap.pip.gr, curr.peaks.gr)
  percent.snps[i] <- length(x)/length(finemap.pip.gr)
  percent.pip[i] <- sum(x$pip)/sum(finemap.pip.gr$pip)
  enrich[i] <- percent.pip[i]/percent.snps[i]
}
```

```{r}
type.names <- gsub(pattern = '_and_', replacement = '\n', x = names(gr.list))
summ.df <- data.frame(percents= c(percent.snps, percent.pip)*100, Type=rep(c('% SNPs','% PIP'), rep(length(type.names), 2)), profile=rep(type.names, 2))
ggplot(summ.df, aes(y=percents, x=profile, fill=Type)) + geom_bar(stat='identity', position='dodge') + ggClean(rotate_axis = F) + ylab('%') + xlab('Peak Cluster') + coord_flip()
```

