<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Heart eQTLs</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">HeartAtlas</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Heart eQTLs</h1>

</div>


<div id="exploration-gtex-enrichment" class="section level3">
<h3>Exploration: GTEx enrichment</h3>
<p>GTEx has eQTL data for n = 495 individuals for heart tissue (left ventricle.) We hypothesize that the cardiomyocyte (CM) regulatory elements are enriched with heart eQTLs. We use Torus to perform this enrichment analysis. We plot the log2 enrichment result below.</p>
<pre class="r"><code>source(&#39;../R/analysis_utils.R&#39;)

enrich.df &lt;- read.delim(&#39;../eQTL_enrich/results/torus_celltypes_combined.enrichment&#39;, header=F, sep=&quot;&quot;)

ggplot(enrich.df, aes(x=V2, y=V1)) + geom_point() + geom_errorbar(aes(xmin=V3, xmax=V4), colour=&quot;black&quot;, width=.1) + ggClean() +
  xlab(&#39;log2 Enrichment&#39;) + ylab(&#39;Cell Type&#39;) + geom_vline(xintercept = 0, col=&#39;red&#39;, linetype = &quot;longdash&quot;) + ggtitle(&#39;Enrichment of ArchR DA Peaks&#39;)</code></pre>
<p><img src="eqtl_enrichment_files/figure-html/gtex1-1.png" width="672" /></p>
</div>
<div id="heart-eqtl-finemapping" class="section level3">
<h3>Heart eQTL Finemapping</h3>
<p>Here we focus on eQTLs that are finemapped with a posterior probability of association &gt; 0.8. There are approximately 2,300 such SNPs.</p>
<pre class="r"><code>finemap.res &lt;- suppressMessages(readr::read_tsv(&#39;../eQTL_enrich/gtex_finemapping/GTEx_v8_finemapping_DAPG/Heart_LV_Finemapping_CS95.txt&#39;, col_names = F))
colnames(finemap.res) &lt;- c(&quot;tissue&quot;,&quot;gene&quot;,&quot;cluster_id&quot;,&quot;cluster_pip&quot;,&quot;variant_id&quot;,&quot;variant_pip&quot;)</code></pre>
<pre class="r"><code>high_pip_eqtls &lt;- unique(finemap.res$variant_id[finemap.res$variant_pip&gt;0.8])
eqtl.gr &lt;- snpIDtoGR(high_pip_eqtls)
eqtl.gr$SNP &lt;- high_pip_eqtls</code></pre>
<pre class="r"><code>rand.snps &lt;- readLines(&#39;../matched_SNPs/eQTL_top_pip_hg19_5batches/snpsnap_match_hg38.txt&#39;)
rand.gr &lt;- StringToGR(rand.snps)
rand.gr &lt;- rand.gr[seqnames(rand.gr) %in% paste0(&quot;chr&quot;,1:22),]
seqlevels(rand.gr) &lt;-  paste0(&quot;chr&quot;,1:22)
rand.gr$SNP &lt;- paste0(seqnames(rand.gr),&#39;_&#39;,start(rand.gr))</code></pre>
<p>Get tissue activity of each eQTL from <a href="https://zenodo.org/record/3727189" class="uri">https://zenodo.org/record/3727189</a> (GTEx v8 paper)</p>
<pre class="r"><code>eqtl.lfsr &lt;- suppressMessages(vroom::vroom(&#39;../eQTL_enrich/broadinstitute-gtex-v8-a014b43/data/Fig6C_all_top.z_lfsr.sig.pruned.txt.gz&#39;))</code></pre>
<p>Keep SNPs with PIP &gt; 0.8 AND they have LFSR data.</p>
<pre class="r"><code>same.snps &lt;- intersect(eqtl.gr$SNP, eqtl.lfsr$variant)
heart.eqtl.lfsr &lt;- eqtl.lfsr[eqtl.lfsr$variant %in% same.snps,]
eqtl.gr &lt;- eqtl.gr[eqtl.gr$SNP %in% same.snps]</code></pre>
<p>Find the mean number of tissues per eQTL.</p>
<pre class="r"><code># all eqtls
tissues.active &lt;- rowSums(eqtl.lfsr[,3:ncol(eqtl.lfsr)] &lt; 0.01, na.rm = T) 
all.eqtl.ntissues &lt;- data.frame(eqtl=eqtl.lfsr$variant, ntissues=tissues.active) %&gt;% group_by(eqtl) %&gt;% summarise(mean_tissues = mean(ntissues))</code></pre>
<pre><code>## `summarise()` ungrouping output (override with `.groups` argument)</code></pre>
<pre class="r"><code>all.eqtl.ntissues &lt;- all.eqtl.ntissues[all.eqtl.ntissues$mean_tissues &gt; 0,]

# heart eqtls
tissues.active &lt;- rowSums(heart.eqtl.lfsr[,3:ncol(heart.eqtl.lfsr)] &lt; 0.01, na.rm = T) 
heart.eqtl.ntissues &lt;- data.frame(eqtl=heart.eqtl.lfsr$variant, ntissues=tissues.active) %&gt;% group_by(eqtl) %&gt;% summarise(mean_tissues = mean(ntissues))</code></pre>
<pre><code>## `summarise()` ungrouping output (override with `.groups` argument)</code></pre>
<pre class="r"><code>heart.eqtl.ntissues &lt;- heart.eqtl.ntissues[heart.eqtl.ntissues$mean_tissues &gt; 0,]</code></pre>
</div>
<div id="tissue-sharing-of-eqtls" class="section level3">
<h3>Tissue Sharing of eQTLs</h3>
<p>Below we plot the frequency of tissues active in causal heart eQTLs. Typically eQTLs have a bimodal distribution with peaks at 1-5 tissues and 45-50 tissues. We see that for heart eQTLs, they are generally shared across tissues.</p>
<pre class="r"><code>betterHist &lt;- function(X){
 bks &lt;- seq(0, 50, 5)
 labs &lt;- sapply(1:(length(bks)-1), function(x){paste0(bks[x]+1,&#39;-&#39;,bks[x+1])})
 bin.count &lt;- table(cut(X$mean_tissues, breaks = bks, labels = labs))
 bin.count &lt;- bin.count/sum(bin.count)
 bin.count.df &lt;- data.frame(count=as.numeric(bin.count), breaks=factor(names(bin.count), levels = labs))
}

all.eqtl.bin &lt;- betterHist(all.eqtl.ntissues)
heart.eqtl.bin &lt;- betterHist(heart.eqtl.ntissues)

result &lt;- rbind(all.eqtl.bin, heart.eqtl.bin)
result$eQTLs &lt;- c(rep(c(&quot;All&quot;,&quot;Heart&quot;), each=nrow(all.eqtl.bin)))</code></pre>
<pre class="r"><code>ggplot(result, aes(x=breaks, y=count, fill=eQTLs)) + geom_bar(stat=&quot;identity&quot;, position = &quot;dodge&quot;, width=0.8) + ggClean() + xlab(&#39;Tissues with LFSR &lt; 0.01&#39;) + ylab(&#39;Proportion of eQTLs&#39;) + ggtitle(&#39;Tissue Sharing of Heart eQTLs&#39;) + theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + scale_fill_manual(values = c(&quot;#D6604D&quot;,&quot;#4393C3&quot;))</code></pre>
<p><img src="eqtl_enrichment_files/figure-html/unnamed-chunk-47-1.png" width="672" /></p>
<p>Add tissue activity to our eQTL genomic ranges object</p>
<pre class="r"><code>eqtl.gr$ntissues.active &lt;- heart.eqtl.ntissues$mean_tissues[match(eqtl.gr$SNP, heart.eqtl.ntissues$eqtl)]
#eqtl.gr.low &lt;- eqtl.gr[which(eqtl.gr$ntissues.active&lt;=10),]
#eqtl.gr.high &lt;- eqtl.gr[which(eqtl.gr$ntissues.active&gt;=30),]</code></pre>
</div>
<div id="distribution-of-heart-eqtls" class="section level3">
<h3>Distribution of heart eQTLs</h3>
<p>Overlap generic genomic annotations with top eQTLs</p>
<pre class="r"><code>disjoint.annots &lt;- readRDS(&#39;../eQTL_enrich/annotations/hg38_disjoint_annotations.gr.rds&#39;)
exon.annots &lt;- disjoint.annots[disjoint.annots$type == &quot;exon&quot;,]
utr.annots &lt;- disjoint.annots[disjoint.annots$type == &quot;UTR&quot;,]
intron.annots &lt;- disjoint.annots[disjoint.annots$type == &quot;intron&quot;,]</code></pre>
<pre class="r"><code>satac &lt;- ArchR::loadArchRProject(path = &#39;../ArchR/ArchR_heart/&#39;, showLogo = F)</code></pre>
<pre><code>## Successfully loaded ArchRProject!</code></pre>
<pre class="r"><code>annots &lt;- as.list(ArchR::geneAnnoHg38)
#annots$promoters &lt;- extendGR(resize(annots$genes, 1, &quot;start&quot;), upstream = 2000, downstream = 100)
#annots$exons &lt;- removeOverlaps(X = annots$exons, to.remove = annots$promoters)</code></pre>
<p>Split ventrical cardiomyocyte peaks into cell-type specific and into shared.</p>
<pre class="r"><code>union.set &lt;- readRDS(&#39;../ArchR/ArchR_heart/PeakCalls/UnionSet.gr.rds&#39;)
union.exons &lt;- union.set[union.set$peakType == &quot;Exonic&quot;, ]

ventCalls &lt;- readRDS(&#39;../ArchR/ArchR_heart/PeakCalls/Vent..CM-reproduciblePeaks.gr.rds&#39;)
markers &lt;- readRDS(&#39;../ArchR/ArchR_heart/PeakCalls/DA_MARKERS_FDRP_1_log2FC_1.rds&#39;)
vent.specific &lt;- markers$`Vent. CM`

# remove Exonic peaks (because we have exon and UTR annotation)
union.set &lt;- removeOverlaps(X = union.set, to.remove = union.exons) 
ventCalls &lt;- removeOverlaps(X = ventCalls, to.remove = union.exons)
vent.specific &lt;- removeOverlaps(X = vent.specific, to.remove = union.exons)

vent.shared &lt;- removeOverlaps(X = ventCalls, to.remove = vent.specific)
non.cm.peaks &lt;- removeOverlaps(X = union.set, to.remove = ventCalls)
non.cm.peaks &lt;- removeOverlaps(X = non.cm.peaks, to.remove = vent.specific)

intron.nonpeak &lt;- removeOverlaps(intron.annots, to.remove = union.set)
intron.nonpeak &lt;- removeOverlaps(intron.nonpeak, to.remove = vent.specific)
intron.nonpeak &lt;- removeOverlaps(intron.nonpeak, to.remove = vent.shared)</code></pre>
<p>Overlap eQTLs with exons, introns, etc.</p>
<pre class="r"><code>peaks.gr.list &lt;- list(&quot;CM Specific&quot; = vent.specific, 
                      &quot;CM Shared&quot; = vent.shared,
                      &quot;Non-CM&quot; = non.cm.peaks, 
                      &quot;Intronic&quot; = intron.nonpeak,
                      &quot;Exon&quot; = exon.annots, 
                      &quot;UTR&quot; = utr.annots)

peaks.eqtls &lt;- join_overlap_list(gr.list = peaks.gr.list, X = eqtl.gr)
peaks.random &lt;- join_overlap_list(gr.list = peaks.gr.list, X = rand.gr)</code></pre>
<pre class="r"><code>snpsIn &lt;- unique(
  unlist(
    sapply(peaks.eqtls, function(x){x$SNP})
  )
)
peaks.eqtls$Unassigned &lt;- eqtl.gr[!(eqtl.gr$SNP %in% snpsIn),]

snpsIn &lt;- unique(
  unlist(
    sapply(peaks.random, function(x){x$SNP})
  )
)
peaks.random$Unassigned &lt;- rand.gr[!(rand.gr$SNP %in% snpsIn),]</code></pre>
<pre class="r"><code>peak.dist.df &lt;- as.data.frame(sapply(peaks.eqtls, FUN = function(x){length(x)/length(eqtl.gr)}))
colnames(peak.dist.df) &lt;- c(&quot;freq&quot;)
peak.dist.df$category &lt;- rownames(peak.dist.df)

random.dist.df &lt;- as.data.frame(sapply(peaks.random, FUN = function(x){length(x)/length(rand.gr)}))
colnames(random.dist.df) &lt;- c(&quot;freq&quot;)
random.dist.df$category &lt;- rownames(random.dist.df)</code></pre>
<pre class="r"><code>peak.set.dist.df &lt;- Reduce(rbind, list(random.dist.df, peak.dist.df))
peak.set.dist.df$SNPs &lt;- c(rep(&quot;Random SNPs&quot;, nrow(random.dist.df)),
                          rep(&quot;eQTLs&quot;, nrow(peak.dist.df)))</code></pre>
<pre class="r"><code>plot.levels &lt;- c(&quot;Exon&quot;,&quot;UTR&quot;,&quot;Intronic&quot;,&quot;CM Specific&quot;,&quot;CM Shared&quot;, &quot;Non-CM&quot;, &quot;Unassigned&quot;)
peak.set.dist.df$category &lt;- factor(peak.set.dist.df$category, 
                                    levels = rev(plot.levels))

#png(&#39;wc_distribution_1.png&#39;, width=1500, height=800, res=150)
ggplot(peak.set.dist.df, aes(y=category, x=freq, fill=SNPs)) + 
  geom_bar(stat=&quot;identity&quot;, position=&quot;dodge&quot;) + ggClean() + xlab(&quot;Wc (prop. of SNPs)&quot;) + ylab(&quot;Category&quot;) + ggtitle(&quot;Distribution of Top Heart eQTLs&quot;) + scale_fill_manual(values = c(&quot;black&quot;,&quot;gray&quot;))  + coord_cartesian(xlim=c(0, 0.4))</code></pre>
<p><img src="eqtl_enrichment_files/figure-html/unnamed-chunk-56-1.png" width="672" /></p>
<pre class="r"><code>#dev.off()</code></pre>
<p>Tissue sharing patterns in scATAC-seq peak sets</p>
<pre class="r"><code>peaks.tissue.sharing &lt;- sapply(peaks.eqtls, function(x){x$ntissues.active}) 
peaks.tissue.sharing$`Union Set` &lt;- NULL
peaks.tissue.sharing &lt;- setNames(unlist(peaks.tissue.sharing, use.names=F),rep(names(peaks.tissue.sharing), lengths(peaks.tissue.sharing)))
peaks.tissue.sharing.df &lt;- data.frame(category=names(peaks.tissue.sharing), ntissues=peaks.tissue.sharing)
peaks.tissue.sharing.df$category &lt;- factor(peaks.tissue.sharing.df$category, levels=rev(plot.levels))
peaks.tissue.sharing.df &lt;- peaks.tissue.sharing.df[!is.na(peaks.tissue.sharing.df$ntissues),]</code></pre>
<pre class="r"><code>pc_shared &lt;- peaks.tissue.sharing.df %&gt;% group_by(category) %&gt;% summarise(p_shared = mean(ntissues &gt; 15, na.rm = T))</code></pre>
<pre><code>## `summarise()` ungrouping output (override with `.groups` argument)</code></pre>
<pre class="r"><code>#png(&quot;p_c_shared.png&quot;, width=1000, height=800, res=150)
ggplot(pc_shared, aes(y = category, x = p_shared)) + geom_bar(stat=&quot;identity&quot;) + ggClean() + xlab(&quot;Pc (prob. sharing)&quot;) + ylab(&quot;Category&quot;)</code></pre>
<p><img src="eqtl_enrichment_files/figure-html/unnamed-chunk-58-1.png" width="672" /></p>
<pre class="r"><code>#dev.off()</code></pre>
<pre class="r"><code>#png(&quot;sharing_violin.png&quot;, width=800, height=1400, res=150)

ggplot(peaks.tissue.sharing.df, aes(y=category, x=ntissues)) + 
  geom_violin(fill=&quot;lightblue&quot;) + 
  geom_boxplot(width=0.1) + 
  ggClean() + LegendOff() + xlab(&quot;Tissues Active (LFSR &lt; 0.01)&quot;) + ylab(&quot;&quot;) +
  scale_x_continuous(breaks=seq(0,50,10))</code></pre>
<p><img src="eqtl_enrichment_files/figure-html/unnamed-chunk-59-1.png" width="672" /></p>
<pre class="r"><code>#dev.off()</code></pre>
<p>Expression patterns</p>
<pre class="r"><code>high_pip_genes &lt;- finemap.res$gene[finemap.res$variant_id %in% peaks.eqtls$`CM Specific`$SNP]
high_pip_genes &lt;- sub(&#39;[.].*&#39;, &#39;&#39;, high_pip_genes)
high_pip_gene_symbol &lt;- ensembldb::select(EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86, keys= high_pip_genes, keytype = &quot;GENEID&quot;, columns = &quot;SYMBOL&quot;)
high_pip_gene_symbol &lt;- high_pip_gene_symbol$SYMBOL</code></pre>
<pre class="r"><code>srna &lt;- readRDS(&#39;../seurat/Heart_RNA_Processed_Combined.rds&#39;)</code></pre>
<pre class="r"><code>get_cell_type_expr &lt;- function(srna, genes){
  rna.mat &lt;- srna@assays$RNA@data
  cm.specific.exp &lt;- rna.mat[rownames(rna.mat) %in% genes,]
  celltypes &lt;- unique(Seurat::Idents(srna))
  mean.exp &lt;- list()
  for(t in celltypes){
    mean.exp[[t]] &lt;- rowMeans(cm.specific.exp[,Seurat::Idents(srna) == t])
  }
  exp.mat &lt;- as.data.frame(mean.exp)
  exp.mat[,&quot;Cardiomyocytes&quot;] &lt;- rowMeans(exp.mat[,c(&quot;Vent..CM&quot;,&quot;Atrial.CM&quot;)])
  exp.mat[,c(&quot;Vent..CM&quot;,&quot;Atrial.CM&quot;)] &lt;- NULL
  exp.mat &lt;- as.matrix(exp.mat)
  exp.mat.z &lt;- sweep(exp.mat - rowMeans(exp.mat), MARGIN = 1, STAT = rowSds(exp.mat), FUN = &#39;/&#39;) 
  
  return(exp.mat.z)
}</code></pre>
<pre class="r"><code>CM.specific.egenes.exp &lt;- get_cell_type_expr(srna, high_pip_gene_symbol)</code></pre>
<pre class="r"><code>high_pip_genes &lt;- finemap.res$gene[finemap.res$variant_id %in% peaks.eqtls$`CM Shared`$SNP]
high_pip_genes &lt;- sub(&#39;[.].*&#39;, &#39;&#39;, high_pip_genes)
high_pip_gene_symbol &lt;- ensembldb::select(EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86, keys= high_pip_genes, keytype = &quot;GENEID&quot;, columns = &quot;SYMBOL&quot;)
high_pip_gene_symbol &lt;- high_pip_gene_symbol$SYMBOL</code></pre>
<pre class="r"><code>CM.shared.egenes.exp &lt;- get_cell_type_expr(srna, high_pip_gene_symbol)</code></pre>
<pre class="r"><code>CM.specific.egenes.exp &lt;- reshape2::melt(CM.specific.egenes.exp)
CM.shared.egenes.exp &lt;- reshape2::melt(CM.shared.egenes.exp)
CM.egenes.exp &lt;- rbind(CM.specific.egenes.exp, CM.shared.egenes.exp)
CM.egenes.exp[,&quot;type&quot;] &lt;- c(rep(&quot;CM Specific eGenes&quot;, nrow(CM.specific.egenes.exp)), rep(&quot;CM Shared eGenes&quot;, nrow(CM.shared.egenes.exp)))</code></pre>
<pre class="r"><code>ggplot(CM.egenes.exp, aes(x=Var2, y=value, fill=type)) + geom_boxplot() + ggClean(rotate_axis = T) + xlab(&quot;&quot;) + ylab(&quot;RNA Expression&quot;)</code></pre>
<p><img src="eqtl_enrichment_files/figure-html/unnamed-chunk-67-1.png" width="672" /></p>
</div>
<div id="afib-gwas-enrichment" class="section level3">
<h3>aFib GWAS Enrichment</h3>
<pre class="r"><code>enrich.df &lt;- readRDS(&#39;../GWAS/Torus_Enrichment_Results_Univariate.df.rds&#39;)
enrich.df$term &lt;- sub(pattern = &#39;_hg19.bed.1&#39;, replacement = &quot;&quot;, x = enrich.df$term)
enrich.df$term &lt;- factor(enrich.df$term, levels=enrich.df$term[order(enrich.df$estimate, decreasing = F)])
enrich.df &lt;- enrich.df[enrich.df$term != &quot;non_CM_eQTLs&quot;,]
enrich.df &lt;- enrich.df[enrich.df$term != &quot;non_CM_peaks&quot;,]

ggplot(enrich.df, aes(y=term, x=estimate)) + geom_point(size=3, color=&#39;darkred&#39;) + geom_errorbar(aes(xmin=low, xmax=high), width=0)  + xlab(&#39;Log2 Enrichment&#39;) + ggtitle(&#39;aFib GWAS Enrichment&#39;) + geom_vline(xintercept = 0, color=&#39;red&#39;, linetype=&#39;dashed&#39;) + theme_bw() + theme(text = element_text(size=16)) + ylab(&#39;&#39;)</code></pre>
<p><img src="eqtl_enrichment_files/figure-html/unnamed-chunk-68-1.png" width="672" /></p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
